{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1><center>Protein multiple sequences alignement</center></h1>\n",
    "\n",
    "author: Rawad Ghostin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "In the first part of the project, we looked at multiple ways of aligning 2 sequences together. Very often, many similar sequences are already known to be homologous. These sequences are expected to have similar regions and functions and it would be better if we could align a given sequence to the whole set instead of a specific member. Aligning to the whole set considerably emphasizes how a given sequence deviated from its origins, rather than how it deviated from another specific sequence. Moreover, the degree and type of residues conservation that are put in evidence enhance our capacity of extracting meaningful biological data from the alignments.<br>\n",
    "There exist multiple methods to study different aspects of the sequences. In this project, we focus on examining the alignment of a set of sequences that belong to the same family, the WW domains. The algorithm is based on the Smith-Waterman implemented in the previous part. This allows us to compare a given protein sequence relatively to a known group of sequences, and consider whether the sequence can/cannot be affiliated to the family.<br>\n",
    "Throughout this report, we will explain the concepts and implement this process. Then, we will be discussing the solutions produced by the algorithm in an attempt to make sense of the observations."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Material and methods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MSA\n",
    "The wide majority of WW proteins notably have a high similarity score.\n",
    "In the first part of the project, we looked at multiple ways of aligning 2 protein sequences together. Yet, comparing **sets of sequences** seems to be more relevant in presenting information about the role played by particular residues at particular locations.\n",
    "Additionally, it emphasizes the degree of residue conservation, how much deviation occurred from the protein origins, instead of another specific protein.\n",
    "Such alignments are called **MSA** (Multiple sequence alignment).\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/msa_example.png\">\n",
    "    <caption> Fig1 : Example of MSA</caption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "#### Progressive algorithm for producing an MSA\n",
    "A wide variety of methods exist to produce MSAs.<br>\n",
    "In our project, we will be using an iterative approach, using the <a href=\"https://www.ebi.ac.uk/Tools/msa/muscle/\">MUSCLE</a> tool.\n",
    "\n",
    "This technique is based on a stochastic method in which the algorithm starts with a random alignment and ameliorates it iteratively until it approaches a good-enough solution.\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/iterative.png\">\n",
    "    <caption> Figure 2 : Diagram explaining the flow of an iterative algorithm</caption>\n",
    "</figure>\n",
    "\n",
    "##### Evaluating the alignment\n",
    "Progressive algorithms operate iteratively by producing new generations until a convenient solution is obtained.\n",
    "We first start with a random population *generation 0* (a set of sequences). For each iteration, the algorithms evaluate the correctness of the current population using the *Sum-of-pairs*, and produces a new generation replacing the current one until a satisfactory evaluation is reached, meaning a certain precision degree is attained.\n",
    "\n",
    "The Sum-of-pairs is defined as follows:\n",
    "$$\n",
    "S(m) = \\sum_{k,l} S(m_{k,l}) \\textrm{     Total sum of pairs (sum of the scores of columns)}\\\\\n",
    "$$\n",
    "\n",
    "Where\n",
    "\n",
    "$$\n",
    "S(m_{k,l}) = \\sum_i s(m^k_i, m^l_i) \\textrm{ score of a column (sum on all possible pairs)}\n",
    "$$\n",
    "\n",
    "And $m^k_i$ is the residue in the sequence $k$ at the position $i$ and $s(m^k_i, m^l_i)$ the score according to the matrix of substitution.<br>\n",
    "A new generation is derived from the current one using the amelioration process which is explained in the next section.\n",
    "\n",
    "##### Amelioration process\n",
    "The amelioration process can be done with multiple techniques (sometimes mixed) like *stochastic hill*, *simulated annealing*, *tabu search* and so forth.\n",
    "For instance, the *SAGA* algorithm ameliorates the alignment using a set of *crossing operators* and *mutation operators*. Crossing operators emulated the genetic crossing done naturally for a new generation of individuals to be born. Mutation operators emulate mutations happening to new the genome of the new generation.\n",
    "\n",
    "For each iteration, operators from both categories are activated. They are selected randomly according to a probability law which is regularly updated base on a reward/penalty system, just like in nature.<br>\n",
    "The operators that are known to produce generations which approach better the required precision are rewarded by increasing the likelihood of their selection in the next iterations.\n",
    "Likewise, poorly performing operators are penalized and have less probability to be picked up in the future.\n",
    "Finally, the algorithm takes into consideration a *best-fit* policy, and only selects the best 50% of individuals to form the new population.\n",
    "\n",
    "Unlike other algorithms which are based on different kind of techniques, e.g *CLUSTALo* for progressive alignment, the *MUSCLE* algorithm has the advantage of being effective for both small and large groups of sequences, thus is our preferred choice.\n",
    "\n",
    "\n",
    "The MSA produced is put in FASTA format, in that way, we can simply reuse our FASTA parser from the previous part of the project.<br>\n",
    "\n",
    "\n",
    "\n",
    "### MSA variables and operations \n",
    "\n",
    "We can already fix variables notations and operations of an MSA. These are useful for computing a profile, a concept explained in fine detail in the next section.\n",
    "\n",
    "- $N_{seq}$ represents the number of sequences aligned in the MSA.<br>\n",
    "- $n_{u,a}$ represents the number of occurence of an amino acid $a$ at the position $u$ of the alignment.<br>\n",
    "- $f_{u,a}$ represents the proportion that an amino acid $a$ holds on a position $u$. <br>\n",
    "It is calculated as follows:<br>\n",
    "$$ \n",
    "f_{u,a}=\\frac{n_{u,a}}{N_{seq}}\n",
    "$$\n",
    "\n",
    "\n",
    "The MSA structure and operations are defined in code as follows, but first we re-define some ADTs and the FASTA parser from the previous part of the project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from math import log10\n",
    "from collections.abc import MutableSequence\n",
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Data structures from the previous part\n",
    "\n",
    "class AminoAcid:\n",
    "    __valid_AA = 'ABCDEFGHIKLMNPQRSTVWXYZ-'     # symbols representing an amino acid\n",
    "\n",
    "    def __init__(self, symbol):\n",
    "        self.__symbol = None\n",
    "        self.symbol = symbol        # call to setter\n",
    "\n",
    "    def __repr__(self):\n",
    "        return 'AA({})'.format(self.symbol)\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.symbol\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        return self.symbol == other.symbol\n",
    "\n",
    "    @property\n",
    "    def symbol(self):\n",
    "        return self.__symbol\n",
    "\n",
    "    @symbol.setter\n",
    "    def symbol(self, symbol):\n",
    "        symbol = symbol.upper()\n",
    "        assert symbol in self.__valid_AA    # asserts that symbol is valid before setting it\n",
    "        self.__symbol = symbol\n",
    " \n",
    "\n",
    "class AASequence(list):\n",
    "    def __init__(self, fasta_header=None):\n",
    "        super().__init__()\n",
    "        self.fasta_header = fasta_header\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\" Returns AASequence in format : 'ACDEFGHIKLMNPQRSTVWY' \"\"\"\n",
    "        return ''.join([str(am_acid) for am_acid in self])\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\" Returns AASequence in format : AASequence(ACDEFGHIKLMNPQRSTVWY) \"\"\"\n",
    "        return 'AASequence({})'.format(str(self))\n",
    "\n",
    "    def fasta_repr(self):\n",
    "        assert self.fasta_header is not None\n",
    "        return self.fasta_header + '\\n' + str(self) + '\\n'\n",
    "\n",
    "    def append(self, symbol):\n",
    "        \"\"\"\n",
    "        Extension to the python list.append.\n",
    "        Converts the str symbol into AminoAcid object before append\n",
    "        \"\"\"\n",
    "        am_acid = AminoAcid(symbol)\n",
    "        super().append(am_acid)\n",
    "\n",
    "    def extend(self, symbol_iterable):\n",
    "        \"\"\"\n",
    "        Overwriting python list.extend\n",
    "        Given an iterable of string symbols, appends them to the sequence\n",
    "        \"\"\"\n",
    "        for symbol in symbol_iterable:\n",
    "            self.append(symbol)     # call to extended append\n",
    "            \n",
    "class Matrix:\n",
    "    def __init__(self, nrows, ncols, init_val=None):\n",
    "        self.nrows = nrows\n",
    "        self.ncols = ncols\n",
    "        self.init_val = init_val if not isinstance(init_val, MutableSequence) else deepcopy(init_val)\n",
    "        self.mat = None\n",
    "        self.reset()        # fill matrix in self.mat\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self.mat[idx]\n",
    "\n",
    "    def __setitem__(self, idx, newrow):\n",
    "        assert len(newrow) == self.nrows\n",
    "        self.mat[idx] = newrow\n",
    "\n",
    "    def __str__(self):\n",
    "        res = ''\n",
    "        for col in self.mat:\n",
    "            res += ' '.join([repr(x) for x in col]) + '\\n'\n",
    "        return res\n",
    "\n",
    "    def __repr__(self):\n",
    "        res = 'Matrix({})'\n",
    "        content = ', '.join([repr(col) for col in self.mat])\n",
    "        return res.format(content)\n",
    "\n",
    "    def reset(self):\n",
    "        self.mat = [[None for j in range(self.ncols)] for i in range(self.nrows)]\n",
    "        for i in range(self.nrows):\n",
    "            for j in range(self.ncols):\n",
    "                if isinstance(self.init_val, MutableSequence):\n",
    "                    self.mat[i][j] = deepcopy(self.init_val)\n",
    "                else:\n",
    "                    self.mat[i][j] = self.init_val\n",
    "\n",
    "    def fillrow(self, i, val):\n",
    "        \"\"\" fills row i with val \"\"\"\n",
    "        for j in range(self.ncols):\n",
    "            self.mat[i][j] = val\n",
    "\n",
    "    def fillcol(self, j, val):\n",
    "        \"\"\" fills col i with val \"\"\"\n",
    "        for i in range(self.nrows):\n",
    "            self.mat[i][j] = val\n",
    "\n",
    "    def getmax(self):\n",
    "        \"\"\" returns value, row, col of maximum in matirx\"\"\"\n",
    "        maxval = 0, 0, 0  # value, i, j\n",
    "        for i in range(self.nrows):\n",
    "            for j in range(self.ncols):\n",
    "                if self.mat[i][j] > maxval[0]:\n",
    "                    maxval = self.mat[i][j], i, j\n",
    "        return maxval\n",
    "    \n",
    "    \n",
    "def parse_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Parses fasta file line by line. todo: remarks: light on memory, for larger files\n",
    "    :param filename:  path to fasta file\n",
    "    :return: list of AASequences\n",
    "    \"\"\"\n",
    "    seq_list = []\n",
    "    curr_seq = None\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        line = f.readline()\n",
    "        while line:\n",
    "            line = line.strip()\n",
    "            # if sequence header, create new instance else continue parsing current one\n",
    "            if line.startswith('>'):\n",
    "                curr_seq = AASequence(fasta_header=line)         # create new QQSequence instance\n",
    "                seq_list.append(curr_seq)       # append reference to the AASequence object\n",
    "            else:\n",
    "                curr_seq.extend(line)\n",
    "            line = f.readline()\n",
    "    return seq_list   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class MSA:\n",
    "    def __init__(self, seq_list):\n",
    "        self.__seq_list = seq_list\n",
    "\n",
    "    @property\n",
    "    def N_seq(self):\n",
    "        \"\"\" Returns number of sequences being aligned \"\"\"\n",
    "        return len(self.__seq_list)\n",
    "\n",
    "    @property\n",
    "    def len(self):\n",
    "        \"\"\" Returns the length of the sequences after being aligned \"\"\"\n",
    "        return len(self.__seq_list[0])\n",
    "\n",
    "    def get_col(self, u):\n",
    "        \"\"\" Helper function, returns the column u in a list format\"\"\"\n",
    "        return [seq[u] for seq in self.__seq_list]\n",
    "\n",
    "    def n(self, u, a):\n",
    "        \"\"\" Returns the number of occurrence of an amino acid a at a position u\"\"\"\n",
    "        a = AminoAcid(a)\n",
    "        return self.get_col(u).count(a)\n",
    "\n",
    "    def f(self, u, a):\n",
    "        \"\"\" Returns the frequency of a at position u\"\"\"\n",
    "        return self.n(u, a) / self.N_seq \n",
    "\n",
    "    def __str__(self):\n",
    "        res = \"[*] MSA:\\nN_seq=%s; len=%s\\n\" % (self.N_seq, self.len)\n",
    "        for seq in self.__seq_list:\n",
    "            res += str(seq) + '\\n'\n",
    "        return res\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Profiles\n",
    "\n",
    "In the first part of the project, we looked at multiple ways of aligning 2 sequences together.\n",
    " \n",
    "Very often, many similar sequences are already known to be homologous.\n",
    "These sequences are expected to have similar regions and functions and it would be better if we could align a given sequence to the whole set instead of a specific member.\n",
    "In fact, aligning to the whole set considerably emphasises how a given sequence deviated from its origins, rather than how it deviated from another specific sequence.\n",
    " \n",
    "To this end, it would be helpful to find a representation of the essential properties of the set of sequences such that the given sequence could be aligned to this representation.\n",
    "Such a representation is a matrix called **profile**, it holds compact information about essential properties of an MSA.\n",
    " \n",
    "### PSSM\n",
    " \n",
    "**PSSM**s (Position-specific-scoring-matrix), which are a type of profile, take in consideration the frequency of each aminoacid in each column and its importance on an evolutionary scale.\n",
    "PSSM, which is a matrix of size $ number\\_of\\_possible\\_aacids * len(MSA)$, contains score representing the probability of finding a given amino acid at a specific position.\n",
    "The score is calculated base on observed and theoritical/statistical data.\n",
    "Hence, in order to generate a PSSM, an MSA is required (a set of aligned sequences) to fix prelimary <u>observed</u> values of the group of sequences.\n",
    "Multiple sources and algorithms exist to perform such alignments.\n",
    "In this project, we'll be using the **MUSCLE** tool.\n",
    "Theoritical/statistical data can be obtained from public databases such as Swissprot.\n",
    "\n",
    "#### Computing a PSSM\n",
    "There are multiple methods of calculating a PSSM. Methods mainly differ on specifications, like taking in consideration the evolution process, which data (observed/theoritical) is given more importance and so forth.\n",
    " \n",
    "In our case, we'll be using the following formula:\n",
    "$$\n",
    "m_{u,a}= log \\frac{q_{u,a}}{p(a)}\n",
    "$$\n",
    "Where:\n",
    "\n",
    "- $q_{u,a}$ represents the probability of occurence of the amino acid $a$ in the position $u$ of the alignment.\n",
    "\n",
    " - $p(a)$ represents the proability of appearance of the amino acid $a$ in nature. It is computed by Swissprot based using statistical anylisis, evaluating their genetic database composition. The probability is dependent on various evolutive factors. For instance, amino acids naturally produced by homo sapiens tend to be more prevalent than other acids which aren't.\n",
    " \n",
    " - $m_{u,a}$ is the score representing the likelyhood of finding the amino acid $a$ at the position $u$ in that alignment.\n",
    "\n",
    "The score is a *log-odds* ratio, which calculated like follows:\n",
    "$$\n",
    "log \\frac{p(event\\_A\\_in\\_specific\\_conditions)}{p(event\\_A\\_in\\_random)}\n",
    "$$\n",
    "\n",
    " \n",
    "### Overcoming lack of data with pseudocounts\n",
    "The log-odds formula for computing the PSSM scores $m_{u,a}$ have the characteristic that if an amino acid $a$ is not observed in a column $u$ the result tends towards $-\\infty$.\n",
    "In consequence, the given to-be-aligned sequence won't be able to align on that position $u$.\n",
    "However, the absence of a specific amino acid in a specific column is most probably due to a lack of data rather than an actual alignment preference. This scheme appears too restrictive.\n",
    " \n",
    "The root of the problem resides in $q_{u,a}$ which can eventually be valued to $0$ on the event of lack of data.\n",
    "A technique used to overcome this issue is adding *pseudocounts*.\n",
    "Pseudocounts represent our knowledge of the system statistically (what we observe), before introducing any new residue. They can be used to *infer*  data in particular positions, thus simulating a complete and rich system with no lack of data.<br>\n",
    "\n",
    "This knowledge can be incorporated in the formula of $q_{u,a}$ as follows:\n",
    "\n",
    "$$\n",
    "q_{u,a} = \\frac{\\alpha \\cdot f(u,a) + \\beta \\cdot p(a)}{\\alpha + \\beta}\n",
    "$$\n",
    "\n",
    "\n",
    "$\\alpha$ and $\\beta$ are simple scaling parameters.<br>\n",
    "$\\alpha$ scales the importance accorded to the frequency of the amino acid $a$ in the column $u$.<br>\n",
    "$\\beta$ scales the importance accorded to the probability of occurence of the amino acid $a$ in nature.<br>\n",
    "More abstractly, these parameters weigh the importance we accord to theoritical and observed data in our system.<br>\n",
    "A division by $ \\alpha+\\beta$ is required to keep the proportionality holding. In fact, the sum of the $q_{u,a}$ for all amino acid a *must* be contained between $0$ and $1$ according to the first and second Kolmogorov axioms of probability.<br>\n",
    "We note that in that manner, having $p(a) \\in ]0,1]$, $q_{u,a}>0$ for all possibilities, thus eliminating the event of having $m_{u,a} = log 0 = -\\infty$.\n",
    "\n",
    "In our project, we set these factors to be:\n",
    "$$\n",
    "\\alpha = N_{seq}-1 \\text{ and } \\beta=\\sqrt{N_{seq}}\n",
    "$$\n",
    "\n",
    "\n",
    "When there is a lot of \"real data\", there is little need to weigh up the pseudocount's importance and $\\beta$ should be low relatively to $N_{seq}$, whereas when we're dealing with a few quantity of data relying on pseudocounts is required, and $\\beta$ should be high relatively to $N_{seq}$. A common technique is setting $\\beta=\\sqrt{N_{seq}}$ as a balanced weight for pseudocounts.\n",
    "This might result of $\\beta$ being too low for a small $N_{seq}$, however for a high $N_{seq}$, this formula approaches, as desired, $q_{u,a} \\approx f_{u,a}$.<br>\n",
    "As for $\\alpha$, setting the paramter to be dependent on $N_{seq}$ seems reasonable: a good parameter would be one that dynamically adapts to the quantity of data available for observation. We note that setting $\\alpha=N_{seq}$ would be a bad choice because $\\alpha.f_{u,a} = N_{seq}\\cdot\\frac{n_{u,a}}{N_{seq}}=n_{u,a}$ yielding a formula where the scaling factor for observed data is suppressed.\n",
    "$$\n",
    "q_{u,a} = \\frac{n_{u,a} + \\beta \\cdot p(a)}{N_{seq} + \\beta}\n",
    "$$\n",
    "\n",
    "Whereas, with $\\alpha=N_{seq}-1$, $\\alpha \\cdot f_{u,a}=(N_{seq}-1)f_{u,a}=n_{u,a}-f_{u,a}$.Understandably, this formula takes into consideration the frequency at which the selected amino acid appears.\n",
    "\n",
    "We define a PSSM structure as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PSSM:\n",
    "    # probability of occurence of aminoacids in nature (in %)\n",
    "    __p = {\n",
    "        'A': 8.28,  'C': 1.36, 'D': 5.45, 'E': 6.76, 'F': 3.86,\n",
    "        'G': 7.09, 'H': 2.27, 'I': 5.99, 'K': 5.85, 'L': 9.67, \n",
    "        'M': 2.43,'N': 4.05, 'P': 4.68, 'Q': 3.94, 'R': 5.53,\n",
    "        'S': 6.50,'T': 5.32, 'V': 6.87, 'W': 1.07, 'Y': 2.91,\n",
    "    }\n",
    "\n",
    "    def __init__(self, msa, alpha, beta, E):\n",
    "        self.alpha = alpha\n",
    "        self.beta = beta\n",
    "        self.msa = msa\n",
    "        # PSSM matrix, initially entirely filled with None (no computation yet)\n",
    "        self.mat = {AA: [None for _ in range(self.msa.len)] for AA in self.aminoacids}\n",
    "        self.compute()\n",
    "\n",
    "    @classmethod\n",
    "    def p(cls, a):\n",
    "        \"\"\" Returns the probability of occurence of aminoacids in nature \"\"\"\n",
    "        a = a.upper()\n",
    "        return cls.__p[a]/100\n",
    "\n",
    "    @property\n",
    "    def aminoacids(self):\n",
    "        \"\"\"returns a list of possible amino acids\"\"\"\n",
    "        return self.__p.keys()\n",
    "\n",
    "    def q(self, u, a):\n",
    "        \"\"\" Returns the probability of occurence of an aminoacid a in the position u\"\"\"\n",
    "        return (self.alpha * self.msa.f(u, a) + self.beta * self.p(a)) / (self.alpha + self.beta)\n",
    "\n",
    "    def m(self, u, a):\n",
    "        \"\"\"\n",
    "        Returns the score of occurence of an aminoacid a in the position u in that alignment.\n",
    "        \"\"\"\n",
    "        if self.p(a) != 0:\n",
    "            return round(log10(self.q(u, a) / self.p(a)), 3)\n",
    "        else:\n",
    "            raise Exception(\"logic error\")\n",
    "\n",
    "    def compute(self):\n",
    "        \"\"\" Compute the PSSM\"\"\"\n",
    "        for a in self.aminoacids:\n",
    "            self.mat[a] = [self.m(u, a) for u in range(self.msa.len)]\n",
    "\n",
    "    def __str__(self):\n",
    "        res = \"[*] PSSM: alpha=%s; beta=%s;\\n\" % (alpha, beta)\n",
    "        for a in self.mat.keys():\n",
    "            res += str(a) + \"|\"\n",
    "            res += ' '.join(map(str, self.mat[a])) + '\\n'\n",
    "        return res\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        return self.mat[str(item)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning to a profile\n",
    "The methods for generating an MSA and PSSM being now established, we can focus on aligning a new sequence to a profile.\n",
    "Aligning a sequence to a profile is a way of adding a new sequence to a known-related group of sequences.\n",
    "\n",
    "The alignment is based on the previously implement **Smith-Waterman** algorithm, a technique for local sequence alignment. \n",
    "\n",
    "A score matrix, $S$, is used to evaluate alignment scores. The size of S is $len(seq\\_to\\_align) * len(aligned\\_msa\\_seq)$.<br>\n",
    "\n",
    "<img src=\"img/profile_allure.png\">\n",
    "\n",
    "\n",
    "For the same reasons as explained in the previous report, the first column and row of the matrix are initialized to $0$ and represent a gap happening.\n",
    "\n",
    "$S$ is computed using dynamic programming techniques and according to the following formula:\n",
    "\n",
    "$$\n",
    "\\begin{equation}\n",
    "    S(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j-1) + PSSM(seq[i-1],j-1) \\\\\n",
    "        S(i-1, j) + PSSM('-',j) \\\\\n",
    "        S(i,j-1) + PSSM('-',j-1)\\\\\n",
    "        0\\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "$$\n",
    "\n",
    "Where:\n",
    "\n",
    "- $S(i,j)$ represents the score of aligning the $i^{th}$ amino acid of the given sequence with the $j^{th}$ column of the PSSM.\n",
    "- $(i,j)$ is the position in the score matrix.\n",
    "- PSSM(a, u) is the PSSM score of having an amino acid $a$ at the column $u$.\n",
    "- Pssm('-',u) represents having a gap at the position $u$ of the alignment.\n",
    "\n",
    "\n",
    "In this formula, we evaluate various possibilities for each step in the alignment. We select the possibility that yields the best score.\n",
    "\n",
    "- $S(i-1,j-1)+PSSM(seq[i-1],j-1)$ represents the score of keeping the current amino acid in the sequence at the current position, taking in consideration the PSSM score of doing so.<br>\n",
    "The position for accessing the PSSM matrix is decremented by 1 on both axes, because of the additional first (gap) row/column into the matrix.\n",
    "\n",
    "- $S(i-1, j) + PSSM('-',j)$ represents the score of keeping the current amino acid in the sequence at the current position, taking in consideration that a gap happened in the MSA at the position $j$, therefore penalizing the score by the according penalty.\n",
    "- $S(i,j-1) + PSSM('-',j-1)$ represents the score of inserting a gap in the sequence, taking into consideration the penalty of a gap at this position.\n",
    "- 0 prevents having negative scores. The reason is that negative scores produce alignments where the high similarity regions are sparsely spread across the alignment. In this case, we prefer to find regions of the protein that are similar in a \"condensed\" way. Therefore, negative values are set to 0 and not considered.\n",
    "\n",
    "\n",
    "In our case gaps are penalised linearily according to the previously defined formula:\n",
    "\n",
    "$$\n",
    "g(n_{gap}) = - E \\cdot n_{gap}\n",
    "$$\n",
    "Where $E=$penalty of gap and $n_{gap}=$number of gaps.<br>\n",
    "$$\n",
    "\n",
    "Therefore, the score matrix formula can be simplified to:\n",
    "$$\n",
    "\\begin{equation}\n",
    "    S(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j-1) - E \\\\\n",
    "        S(i-1, j) - E \\\\\n",
    "        S(i,j-1) - E\\\\\n",
    "        0\\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "$$\n",
    "\n",
    "\n",
    "Similarly to Smith-Waterman, the algorithm first computes the score matrix, while logging the path taken in a `PathMatrixMSA` (a modified version of the previous `PathMatrix`).\n",
    "Then, having computed the score matrix, the algorithm proceeds to generate a solution with the function `next_sequence`.\n",
    "Solutions are produced by retracing the path with the `tracepath` function while marking cells into the `blacklist`.\n",
    "\n",
    "\n",
    "The structures defining `PathMatrixMSA` and a solution are defined as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PathMatrixMSA(Matrix):\n",
    "    \"\"\" Structure to log the path taken while calculating the score matrix \"\"\"\n",
    "    DIAG, LEFT, TOP = 0, 1, 2\n",
    "\n",
    "    def __init__(self, seq, pssm):\n",
    "        self.pssm = pssm\n",
    "        self.seq = seq\n",
    "        self.init_val = [False, False, False]\n",
    "        super().__init__(ncols=self.pssm.msa.len+1, nrows=len(seq)+1, init_val=self.init_val)\n",
    "\n",
    "    def isdiag(self, i, j):\n",
    "        return self.mat[i][j][self.DIAG]\n",
    "\n",
    "    def isleft(self, i, j):\n",
    "        return self.mat[i][j][self.LEFT]\n",
    "\n",
    "    def istop(self, i, j):\n",
    "        return self.mat[i][j][self.TOP]\n",
    "\n",
    "    def isnone(self, i, j):\n",
    "        return not any(self.mat[i][j])      # [False, False, False]\n",
    "\n",
    "    def setdiag(self, i, j):\n",
    "        self.mat[i][j][self.DIAG] = True\n",
    "\n",
    "    def setleft(self, i, j):\n",
    "        self.mat[i][j][self.LEFT] = True\n",
    "\n",
    "    def settop(self, i, j):\n",
    "        self.mat[i][j][self.TOP] = True\n",
    "\n",
    "    def erase(self, i, j):\n",
    "        \"\"\" reset cell to default [false, false, false] \"\"\"\n",
    "        self.mat[i][j] = deepcopy([False, False, False])\n",
    "\n",
    "    def isdivergent(self, i, j):\n",
    "        \"\"\" checks if multiple origins possible to the cell\"\"\"\n",
    "        return self.mat[i][j].count(True) > 1\n",
    "\n",
    "    def fillrow(self, i, diag, left, top):\n",
    "        for j in range(self.ncols):\n",
    "            # diag, left and top are booleans\n",
    "            x = [None, None, None]\n",
    "            x[self.DIAG] = diag\n",
    "            x[self.TOP] = top\n",
    "            x[self.LEFT] = left\n",
    "            self.mat[i][j] = deepcopy(x)\n",
    "\n",
    "    def fillcol(self, j, diag, left, top):\n",
    "        for i in range(self.nrows):\n",
    "            # diag, left and top are booleans\n",
    "            x = [None, None, None]\n",
    "            x[self.DIAG] = diag\n",
    "            x[self.TOP] = top\n",
    "            x[self.LEFT] = left\n",
    "            self.mat[i][j] = deepcopy(x)\n",
    "            \n",
    "\n",
    "class SolutionBasic:\n",
    "    def __init__(self, s2, startpos, endpos):\n",
    "        self.s2 = s2\n",
    "        self.startpos = startpos\n",
    "        self.endpos = endpos\n",
    "\n",
    "    def __str__(self):\n",
    "        res = self.s2 + '\\n'\n",
    "        res += \"Position : %d - %d\\n\" % (self.startpos, self.endpos)\n",
    "        res += 'Length brute : %d\\n' % len(self.s2)\n",
    "        res += 'Length net : %d\\n' % len([x for x in self.s2 if x!='-'])\n",
    "        return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The algorithm for aligning sequences to a profile is defined below as `PSSMAligner`.\n",
    "The structure takes in parameter a PSSM, a sequence to be aligned to the PSSM, the penalty $E$.\n",
    "An additional parameter $l$ designates how many solutions need to be computed.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PSSMAligner:\n",
    "    def __init__(self, seq, pssm, E, l):\n",
    "        self.seq = seq\n",
    "        self.pssm = pssm\n",
    "        self.E = E\n",
    "        self.l = l\n",
    "\n",
    "        self.S_mat = Matrix(nrows=len(self.seq)+1, ncols=pssm.msa.len +1)\n",
    "        # S matrix initialization\n",
    "        self.S_mat.fillrow(i=0, val=0)\n",
    "        self.S_mat.fillcol(j=0, val=0)\n",
    "\n",
    "        self.path_mat = PathMatrixMSA(seq=self.seq, pssm=self.pssm)\n",
    "        self.blacklist = Matrix(nrows=len(self.seq)+1, ncols=pssm.msa.len +1, init_val=False)\n",
    "        self.solutions = []  # Holds structures of type SolutionBasic\n",
    "\n",
    "    \n",
    "    def display(self):\n",
    "        \"\"\" display solutions \"\"\"\n",
    "        hr = '-' * 20   # horizontal line\n",
    "        print(hr, '[*] Solution(s)', hr, sep='\\n')\n",
    "        print('- seq: %s' % self.seq, '- E=%.2f' % self.E, sep='\\n')\n",
    "        print('- Seq len : %s' % len(self.seq))\n",
    "        print('- PSSM dim : %s x %s' % (len(self.pssm.mat.keys()), self.pssm.msa.len))\n",
    "        print('- S dim : %s x %s' % (self.S_mat.nrows, self.S_mat.ncols))\n",
    "        print('- # solutions : ', len(self.solutions),'\\n')\n",
    "        print('\\n',self.pssm.msa,'\\n')\n",
    "        for sol in self.solutions:\n",
    "            print(sol, end='\\n\\n')\n",
    "        print(hr)\n",
    "\n",
    "    def S(self, i, j):\n",
    "        \"\"\" Computes S(i,j) according to the formula\"\"\"\n",
    "        if self.S_mat[i][j] is None:   # dynamic programing\n",
    "            aa = self.seq[i-1]\n",
    "            diag_score = self.S(i - 1, j - 1) + self.pssm[aa][j-1]\n",
    "            top_score = self.S(i-1, j) + self.E\n",
    "            left_score = self.S(i, j-1) + self.E\n",
    "            self.S_mat[i][j] = max(0, diag_score, top_score, left_score)\n",
    "\n",
    "            # log path, diagonal is favoured\n",
    "            if self.S_mat[i][j] == diag_score:\n",
    "                self.path_mat.setdiag(i, j)\n",
    "            elif self.S_mat[i][j] == top_score:\n",
    "                self.path_mat.settop(i, j)\n",
    "            elif self.S_mat[i][j] == left_score:\n",
    "                self.path_mat.setleft(i, j)\n",
    "        return self.S_mat[i][j]\n",
    "    \n",
    "    def setzero(self, i, j):\n",
    "        \"\"\" mark cell as already treated, and set values to 0\"\"\"\n",
    "        self.blacklist[i][j] = True\n",
    "        self.S_mat[i][j] = 0\n",
    "        self.path_mat.erase(i, j)\n",
    "\n",
    "    def tracepath(self, i, j):\n",
    "        \"\"\" Given initial position (i,j) find sequence alignement and the path taken\n",
    "        according to the SmithWaterman algorithm\"\"\"\n",
    "        s2 = ''\n",
    "        path = []\n",
    "\n",
    "        while not self.S_mat[i][j] == 0:\n",
    "            path.append((i,j))\n",
    "            if self.path_mat.isdiag(i, j):\n",
    "                s2 += str(self.seq[i - 1])\n",
    "                i = i-1\n",
    "                j = j-1\n",
    "\n",
    "            elif self.path_mat.isleft(i, j):\n",
    "                s2 += '-'\n",
    "                i = i\n",
    "                j = j-1\n",
    "\n",
    "            elif self.path_mat.istop(i, j):\n",
    "                s2 += str(self.seq[i - 1])\n",
    "                i = i-1\n",
    "                j = j\n",
    "            else:\n",
    "                raise Exception('Logic error')\n",
    "        return s2, path\n",
    "\n",
    "    def next_solution(self):\n",
    "        \"\"\"\n",
    "        Generates aligned sequences\n",
    "        \"\"\"\n",
    "        # get max and calculate path\n",
    "        value, maxi, maxj = self.S_mat.getmax()\n",
    "        s, path = self.tracepath(i=maxi, j=maxj)\n",
    "        s = s[::-1]\n",
    "\n",
    "        # mark and set to 0\n",
    "        for i, j in path:\n",
    "            self.setzero(i, j)\n",
    "\n",
    "        # erase submatrix starting from last position before reaching the 0 to absolute bottom right position\n",
    "        init_i, init_j = path[-1]\n",
    "        for i in range(init_i, self.S_mat.nrows):\n",
    "            for j in range(init_j, self.S_mat.ncols):\n",
    "                if not self.blacklist[i][j]:\n",
    "                    self.S_mat[i][j] = None\n",
    "\n",
    "        # recalculate matrix\n",
    "        for i in range(init_i, self.S_mat.nrows):\n",
    "            for j in range(init_j, self.S_mat.ncols):\n",
    "                if not self.blacklist[i][j]:\n",
    "                    self.S_mat[i][j] = self.S(i, j)\n",
    "\n",
    "        # start_i and end_i\n",
    "        return SolutionBasic(s2=s, startpos=path[-1][0], endpos=maxi)\n",
    "\n",
    "    def run(self):\n",
    "        # fill the score matrix\n",
    "        assert self.S_mat.nrows == len(self.seq)+1 and self.S_mat.ncols == self.pssm.msa.len+1\n",
    "        for row in range(1, self.S_mat.nrows):\n",
    "            for col in range(1, self.S_mat.ncols):\n",
    "                self.S_mat[row][col] = self.S(row, col)\n",
    "\n",
    "        # compute solutions\n",
    "        for _ in range(self.l):\n",
    "            self.solutions.append(self.next_solution())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reducing the dataset\n",
    "We get our dataset of sequences from *SMART*, a public genetic database. For continuity, our study will focus on WW protein sequences that belong to homo sapiens (humans). The data is in FASTA format and is stored in `to-be-aligned.fasta`.\n",
    "\n",
    "The dataset contains 320 different (but similar) WW sequences.\n",
    "A large number of these sequences are highly similar, thus, we suppose that reducing the dataset can help us approximate the end results, in a more optimized and confined way.\n",
    " \n",
    "In an attempt to filter out our dataset, we consider sequences with a similarity greater than 60% to be duplicated.\n",
    "Therefore, we want to generate a reduced dataset where all pair of sequences have a similarity less than $THRESHOLD=60\\%$.\n",
    " \n",
    "We use a heuristic-based algorithm, it works by linearly grouping sequences in a way that all pairs of sequences of a group have a similarity rate less than 60%\n",
    "Then, the group holding the more sequences will be selected as the new reduced dataset.\n",
    "We choose the largest group so our results would be based on more data, statically making them more accurate.\n",
    " \n",
    "More specifically, in our case, we compare 2 sequences by aligning them with the Needleman-Wunsch algorithm developed in the previous part.\n",
    "The evolutionary scores are computed with an insertion gap of $I=12$ and an extension gap of $E=2$.\n",
    "The **BLOSUM62** substitution matrix was used for these alignments. Although other substitution matrices could have been used, we observed after some testing that all matrices produce (more or less) good enough results. These tests and results are deemed out of scope for this report.\n",
    "\n",
    "\n",
    "The resulting dataset is stored in `to-be-aligned-reduced.fasta`, it contains 15 sequences.\n",
    " \n",
    "For optimization purposes, we used threading techniques (*thread pool*), to compute the comparison with all sequences of a group in parallel.\n",
    "A thread pool is a fixed number of threads complementarily working on a shared queue of tasks.\n",
    " \n",
    "The exact script used for reducing the dataset, *reduce_dataset.py*, yields the file 'to-be-aligned-reduced.fasta' and is joint to the notebook.\n",
    "The script contains some code overhead required to function (ADTs, parsers, Needleman-Wunsch implementation from project1). The less-useful code is written inside the tags \"#### OVERHEAD - start/end\"."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results and discussions\n",
    "We can start by defining some constants, mainly file names and the `PENALTY` value.\n",
    "Having gathered the sequences from *MUSCLE*, we parse the generated FASTA files in order to construct our structures, notably the MSAs and PSSMs. The alignment parameters $\\alpha$ and $\\beta$ are computed according to the number of sequences ($N_seq$) in the MSA.\n",
    "\n",
    "The gap penalty is set to $-0.20$. The value is set heuristically based on trials. Stricter penalties are not satisfactory because they (almost) don't produce gaps in the alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MSA files\n",
    "MSA_FILE = 'rsrc/msaresults-muscle.fasta'\n",
    "MSA_REDUCED_FILE = 'rsrc/msaresults-reduced-muscle.fasta'\n",
    "# protein sequences to be aligned\n",
    "PROT_FILE = 'rsrc/protein-sequences.fasta'\n",
    "# settings\n",
    "PENALTY = -0.20\n",
    "\n",
    "# parsed fastas\n",
    "seq_list = parse_fasta_file(PROT_FILE)\n",
    "msa_seq_list = parse_fasta_file(MSA_FILE)\n",
    "msa_reduced_seq_list = parse_fasta_file(MSA_REDUCED_FILE)\n",
    "\n",
    "# msa and pssm\n",
    "msa = MSA(msa_seq_list)\n",
    "alpha = msa.N_seq-1\n",
    "beta = msa.N_seq**0.5\n",
    "pssm = PSSM(msa=msa, alpha=alpha, beta=beta, E=PENALTY)\n",
    "\n",
    "# msa_reduced and pssm_reduced\n",
    "msa_reduced = MSA(msa_reduced_seq_list)\n",
    "alpha_reduced = msa_reduced.N_seq-1\n",
    "beta_reduced = msa_reduced.N_seq**0.5 \n",
    "pssm_reduced = PSSM(msa=msa_reduced, alpha=alpha_reduced, beta=beta_reduced, E=PENALTY)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PSSM evaluation\n",
    "The scores of a PSSM are effective for obtaining and representing alignments, nonetheless, they do not conveniently *display* the conservation or preference of amino acids at particular locations. This information is significant because it shows the essential *functional* locations of the family. **Logos** are a graphical representation effective enough to identify these fundamental/primary residues.\n",
    "\n",
    "A logo is computed by calculating the quantity of information of each column $u$ in the sequence.\n",
    "$$\n",
    "I_u = log_2 20 - H_u \\\\\n",
    "$$\n",
    "\n",
    "where $H_u$ is the incertitude, for which the formula is:\n",
    "$$\n",
    "H_u = - \\sum f_{u,a}log_2(f_{u,a})\n",
    "$$\n",
    "\n",
    "The contribution of each residue is  $ f_{u,a} \\cdot I_u$.\n",
    "The in-depth explanation of these formulae is deemed out of scope for our project.\n",
    "For more details see C.Shannon's pioneering work of information theory: <a href=\"http://www.math.harvard.edu/~ctm/home/text/others/shannon/entropy/entropy.pdf\">A Mathematical Theory of Communication</a>\n",
    "\n",
    "For convenience, we compute the logos using <a href=\"http://weblogo.threeplusone.com\">Weblogo</a>, an online tool.\n",
    " <figure>\n",
    " <figcaption>Weblogo generated based on the PSSM of the MSA of the 320 sequences</figcaption>\n",
    "  <img src=\"img/logo_muscle.png\">\n",
    "  \n",
    "</figure> \n",
    "<br><br>\n",
    " <figure>\n",
    "   <figcaption>Weblogo generated based on the PSSM of the MSA of the reduced dataset (15 sequences)</figcaption>\n",
    "  <img src=\"img/logo_muscle_reduced.png\">\n",
    "</figure> \n",
    "<br>\n",
    "\n",
    "The \"accepted\" logo of the WW domains available on <a href=\"https://pfam.xfam.org\"></a>.\n",
    " <figure>\n",
    "  <img src=\"img/logo_HMM.png\">\n",
    "</figure> \n",
    "\n",
    "As we can observe, all 3 figures are highly similar. There appear to be 2 sites where the $W$ amino acid is substantially dominant and conservated, at position 6 and 72.\n",
    "Therefore we can deduce that this characteristic is highly essential in the function of the protein, also giving the name to the family, **WW** domains.<br>\n",
    "Other areas are also quite conservated, like the amino acid P at position 76.\n",
    "Unlike the HMM-logo, we can notice a gap of information from position 20-22 to 65 in both of our logos, the reasons are the gaps at these positions in our alignments.\n",
    "\n",
    "Besides, we can deduce that the PSSM we computed based on the MSAs are valid. <br>\n",
    "As a consequence, this also verifies that our reduction of the dataset indeed produces approximated results correctly, as expected.\n",
    "\n",
    "Additionally, we note that the PSSM of the reduced dataset presents less \"pronounced\" information due to a more significant incertitude in information as a consequence of having fewer quantity of data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment evaluation\n",
    "\n",
    "In order to test our algorithm in various situations, we will first evaluate the alignment of sequences in the `protein-sequences.fasta` with both MSAs (full and reduced).\n",
    "Moreover, we will be discussing the results and try to make sense of it by proposing hypotheses \n",
    "\n",
    "## Alignment based on the complete dataset\n",
    "\n",
    "In the following, we align our sequences from `protein-sequences.fasta` with the PSSM of the complete dataset (320 sequences).\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for seq in seq_list:\n",
    "    pa = PSSMAligner(seq=seq, pssm=pssm, E=PENALTY, l=4)\n",
    "    pa.run()\n",
    "    pa.display()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment based on the reduced dataset\n",
    "In the following, we align our sequences from `protein-sequences.fasta` with the PSSM of the reduced dataset (320 sequences).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for seq in seq_list:\n",
    "    pa = PSSMAligner(seq=seq, pssm=pssm_reduced, E=PENALTY, l=4)\n",
    "    pa.run()\n",
    "    pa.display()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison of solutions\n",
    "\n",
    "The comparison of both alignments is presented as follows.<br>\n",
    "As we can observe, both alignments do not produce exactly the same results.\n",
    "Although at least one solution is common between the 2 parties, the alignment produced with the reduced dataset tend to diverge after some solutions\n",
    "The alignments produced with the reduced dataset tend to diverge from the ones produced with the complete dataset, after the first or second solution.<br>\n",
    "However, we are still able to identify similar chunks from both ends.<br>\n",
    "For instance, for the sequence 0, we can observe similarities between the third and the fourth solution of the complete and reduced dataset respectively.\n",
    "\n",
    "\n",
    "LHAQPIIS<span style=\"background-color:rgb(51, 153, 254)\">IRVWGV-GR-D--SGRERDFAYVARD--KLT</span> <br>and<br>\n",
    "<span style=\"background-color:rgb(51, 153, 254)\">I-R-VW--GV-GRDSGRERDFAYVARDKLT</span>QMLK--C-H\n",
    "\n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;}\n",
    ".tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}\n",
    ".tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}\n",
    ".tg .tg-l6li{font-size:10px;border-color:inherit;text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "  <tr>\n",
    "    <th class=\"tg-l6li\">protein ID</th>\n",
    "    <th class=\"tg-l6li\">alignmets<br></th>\n",
    "    <th class=\"tg-l6li\">alignments reduced<br></th>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-l6li\">O00213</td>\n",
    "    <td class=\"tg-l6li\">DLPAGW-MRVQD--TSG-TYYW-HIPT-------------------------GTTQWEPPGR<br><br>SL--GWVEM--TEEELAP-GRSSV--AVN-N<br><br>LHAQPIISIRVWGV-GR-D--SGRERDFAYVARD--KLT<br><br><br>PLPQEE-EKLPPR-NTNP--GIKC-FAV<br></td>\n",
    "    <td class=\"tg-l6li\">DLPAGW-MRVQD--TSG-TYYW-HIPT-------------------GTTQWEPPGR<br><br>-TT----QW<br><br>WATLSQG-SPSYGSPEDTDSF-WNPNAFET-DS--DLPA-GWMR--V-----QDTSGTYYW-HIPT<br><br>I-R-VW--GV-GRDSGRERDFAYVARDKLTQMLK--C-H<br></td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-l6li\">P46937</td>\n",
    "    <td class=\"tg-l6li\">PLPDGWEQAMTQ-DGEIYYINHKN--------------------------KTTSWLDPRL<br><br>PLPAGWEMAKT-SSGQR-YFLNHID-------------------------QTTTWQDPR<br><br>PVSSPG-MSQELRTM-TTNSSDPFLNSG-TY---HSRDEST<br><br>PRTPDDFLN-SVDEM--D--TGDT--IN-QST<br></td>\n",
    "    <td class=\"tg-l6li\">PLPDGWEQAM-T-QDGEIYYINH-------K-N-----------KTTSWLDPR<br><br>PLPAGWE-M--AKTSSGQR-YF-LNHIDQT---------------------TTWQDPR<br><br>QNPVSSPG-MSQELRTMTTNSSDPFLNSG-T-YHSRD-EST<br><br>DFLNS--VDEM-D--TG-DT--IN-QST<br></td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-l6li\">O75554</td>\n",
    "    <td class=\"tg-l6li\">DL-ISGASQWEKPEGFQGDLKKTAVKTVWVEGLSEDG-FTYYYNTET-------------------------GESRWEKPDD<br><br>D-PSKGRWVEGIT-SEGYHYYYD-LIS-------------------------G-ASQWEKP<br><br>--YHYY<br><br>WVEGLSEDGFTYYYNTE-TGESRWEKP-DDFI-PH-T<br></td>\n",
    "    <td class=\"tg-l6li\">TVWVEGLS-EDG-FTYYYN--T------E--T---------GE-SRWEKPDD<br><br>D-PSKGRWVEGIT-SEG-YHYYYDL-IS-------------------G-ASQWEKP<br><br>NPYGEWQEIKQEVESHEEVDLELPSTENE--YV-STS<br><br>-FTYY<br></td>\n",
    "  </tr>\n",
    "</table>\n",
    "\n",
    "Therefore, we can deduce that reducing the dataset produce approximated solutions to the original dataset.. \n",
    "It is reasonable to suppose that the complete dataset produces more accurate solutions. In fact, the full dataset contains more entries which gives a more authentic significance to $f_{u,a}$ while calculating the PSSM. Additionally, since the reduction algorithm operates heuristically, its correctness is not proven for all cases, and the results depend in the order in which the sequences have been read.<br>\n",
    "On the other hand, the alignments based on the complete dataset take in consideration **all** possible WW protein sequences, therefore all relevant information is guaranteed to be considered during the computation.\n",
    "\n",
    "\n",
    "## Comparison using Uniprot\n",
    "For integrity, we compare our alignments to the online database <a href=\"https://www.uniprot.org\">Uniprot</a>.\n",
    "Uniprot offers varying detailed genetical and biological information about proteins.\n",
    "Consulting the *family & domains* sections, we can learn what protein domains are present in the seqence and at which postion the domain resides. This information is quite useful to validate the exactitude of our alignments.\n",
    "\n",
    "Below is a table comparing our alignments'postions with uniprot.\n",
    "\n",
    "<style type=\"text/css\">\n",
    ".tg  {border-collapse:collapse;border-spacing:0;}\n",
    ".tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}\n",
    ".tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}\n",
    ".tg .tg-lboi{border-color:inherit;text-align:left;vertical-align:middle}\n",
    ".tg .tg-0lax{text-align:left;vertical-align:top}\n",
    ".tg .tg-0pky{border-color:inherit;text-align:left;vertical-align:top}\n",
    "</style>\n",
    "<table class=\"tg\">\n",
    "  <tr>\n",
    "    <th class=\"tg-lboi\">protein ID</th>\n",
    "    <th class=\"tg-lboi\">Position of WW (Uniprot)</th>\n",
    "    <th class=\"tg-lboi\">Position of WW full dataset</th>\n",
    "    <th class=\"tg-0lax\">Position of WW reduced dataset</th>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-lboi\">O00213</td>\n",
    "    <td class=\"tg-lboi\">253 – 285</td>\n",
    "    <td class=\"tg-lboi\">254-285</td>\n",
    "    <td class=\"tg-0lax\">254-285</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-lboi\">P46937</td>\n",
    "    <td class=\"tg-lboi\">171 – 204&nbsp;&nbsp;&nbsp;(WW1)<br>230 – 263 (WW2)<br></td>\n",
    "    <td class=\"tg-lboi\">172-203<br>231-263</td>\n",
    "    <td class=\"tg-0lax\">172-203<br>231-264</td>\n",
    "  </tr>\n",
    "  <tr>\n",
    "    <td class=\"tg-0pky\">O75554</td>\n",
    "    <td class=\"tg-0pky\">122 – 155&nbsp;&nbsp;(WW1)<br>163 – 196&nbsp;&nbsp;(WW2)<br></td>\n",
    "    <td class=\"tg-0pky\">122-142<br>169-183<br></td>\n",
    "    <td class=\"tg-0lax\">122-141<br>167-182<br></td>\n",
    "  </tr>\n",
    "</table>\n",
    "\n",
    "As we can observe, in the protein O00213, the WW domain resides at the position 253-285 which is equal (more or less) to the positions produced by our results, 254-285 for both the full and reduced dataset.<br>\n",
    "In the protein P46937, there exists 2 WW domains at position 171-204 and 230-263 respectively, which is equal in  both datasets that produce the position 172-203 for WW1 and 231-263/264 for WW2.<br>\n",
    "In the protein O75554, there exists 2 WW domains at position 122-155 and 163-196 respectively, which is equal in  both datasets that produce the position 122-141/142 for WW1 and 167/169-182/183 for WW2.<br>\n",
    "We note that for P46937 and O75554, since there are 2 WW domains it is required to have $l\\ge2$. The solutions produced in this instance (and that are matching the table above) are the first 2 solutions.\n",
    "\n",
    "We can deduce that both of our alignments produce conviently accurate results compared to Uniport and therefore, are correct. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "In conclusion, we examine in this report protein sequence alignment against an MSA using a PSSM. Key principles are explained and illustrated throughout the report.\n",
    "\n",
    "The algorithms are implemented and tested in an attempt to make sense of different concepts. The results are analyzed, compared to external tools for verification, and discussed accordingly in an attempt to produce hypotheses about the outcome."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference\n",
    "- http://smart.embl.de\n",
    "- https://www.uniprot.org/\n",
    "- https://www.ebi.ac.uk/Tools/msa/muscle/\n",
    "- https://www.comp.nus.edu.sg/~ksung/algo_in_bioinfo/slides/Ch6_MSA.pdf\n",
    "- http://www.math.harvard.edu/~ctm/home/text/others/shannon/entropy/entropy.pdf\n",
    "- http://weblogo.threeplusone.com\n",
    "- Introduction to Bioinformatics, M.Zvelebil and J.OBaum\n",
    "- https://uv.ulb.ac.be/pluginfile.php/1489420/mod_resource/content/1/L4%20Alignement%20de%20plusieurs%20sequences.pdf\n",
    "- http://pfam.xfam.org\n",
    "- Profile analysis: Detection of distantly related proteins , M.Gribskov, A.D.McLachlan, D.Eisenberg\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
