{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h1><center>Needleman-Wunsch and Smith-Waterman <br>for global and local alignments</center></h1>\n",
    "\n",
    "Author: Rawad Ghostin"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "The invention of DNA sequencing in the 1970s has yielded an explosion of genetic information.\n",
    "Millions of nucleotide and amino acid sequences representing the genome of many organisms are stored in large databases, such as GenBank or UniprotKB.\n",
    "Although we are able to digitally read protein sequences, little is known about the function and structure of their different regions.\n",
    "Producing biological information from genetical information is a main goal in bioinformatics.\n",
    "\n",
    "There exist many methods to study different aspects of the sequences. In this project, we only focus on examining the alignment of amino acid sequences, allowing us to compare two sequences, compute their similarity ratio and consider their homology.<br>\n",
    "We will explain and implement different types of sequence alignments through the Needleman-Wunsch and Smith-Waterman algorithms."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Material and Methods\n",
    "\n",
    "The genome is the complete set of genes or genetic material present in a cell or organism.\n",
    "To study the genome we need to sequence the genetic material, translating each molecule to a corresponding character representation.\n",
    "\n",
    "The genetic material can take 2 figures :\n",
    " - DNA: composed of a chain of nucleotide molecules\n",
    " - Protein: composed of a chain of amino acid molecules\n",
    " \n",
    "Sequence comparisons are usually done on protein sequences because they show up similarities more evidently.\n",
    "The main reason for this acuteness is that there are 20 amino acids compared to only 4 nucleotides, so a DNA sequence provides on average a lower quantity of information (entropy) than a protein sequence.\n",
    "\n",
    "Designing a technique for comparing sequences is not straight-forward. In addition to the substantial amount of information, many issues need to be taken into consideration.<br>\n",
    "There are multiple ways a protein sequence can be altered during the evolution:\n",
    "Mutations over multiple generations can result in a substantially altered sequence in comparison to the original ancestral gene.\n",
    "\n",
    "Mutations can come in many forms:\n",
    "- Amino acids molecules can be substituted by others \n",
    "- Insertions and deletions of new molecules (*gaps or indels*) can occur, varying the length of the new sequence\n",
    "\n",
    "The \"noise\" introduced by mutations makes studying similarities a difficult task.\n",
    "Therefore, we try to line up the 2 sequences we are comparing in such a way that common ancestral genes are visibly superposed.\n",
    "\n",
    "Thus, \"Sequence alignment\" is trying to identify regions of two sequences to maximize their similarity based on a score function.\n",
    "\n",
    "There are 2 categories of sequence alignment we will implement in our project, global and local alignments.\n",
    "Both will be discussed in great detail in farther sections.\n",
    "\n",
    "We will first define some structures to represent the protein we'll be working on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections.abc import MutableSequence\n",
    "from copy import deepcopy\n",
    "from itertools import combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class AminoAcid:\n",
    "    __valid_AA = 'ABCDEFGHIKLMNPQRSTVWXYZ'     # symbols representing amino acids\n",
    "\n",
    "    def __init__(self, symbol):\n",
    "        self.__symbol = None\n",
    "        self.symbol = symbol        # call to setter\n",
    "\n",
    "    def __repr__(self):\n",
    "        return 'AA({})'.format(self.symbol)\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.symbol\n",
    "\n",
    "    @property\n",
    "    def symbol(self):\n",
    "        return self.__symbol\n",
    "\n",
    "    @symbol.setter\n",
    "    def symbol(self, symbol):\n",
    "        assert symbol in self.__valid_AA    # asserts that symbol is valid before setting it\n",
    "        self.__symbol = symbol\n",
    "\n",
    "\n",
    "class AASequence(list):\n",
    "    \"\"\"Amino acid chain\"\"\"\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "    def __str__(self):\n",
    "        \"\"\" Returns AASequence in format : 'ACDEFGHIKLMNPQRSTVWY' \"\"\"\n",
    "        return ''.join([str(am_acid) for am_acid in self])\n",
    "\n",
    "    def __repr__(self):\n",
    "        \"\"\" Returns AASequence in format : AASequence(ACDEFGHIKLMNPQRSTVWY) \"\"\"\n",
    "        return 'AASequence({})'.format(str(self))\n",
    "\n",
    "    def append(self, symbol):\n",
    "        \"\"\"\n",
    "        Extension to the python list.append.\n",
    "        Converts the str symbol into AminoAcid object before append\n",
    "        \"\"\"\n",
    "        am_acid = AminoAcid(symbol)\n",
    "        super().append(am_acid)\n",
    "\n",
    "    def extend(self, symbol_iterable):\n",
    "        \"\"\"\n",
    "        Overwriting python list.extend\n",
    "        Given an iterable of string symbols, appends them to the sequence\n",
    "        \"\"\"\n",
    "        for symbol in symbol_iterable:\n",
    "            self.append(symbol)     # call to extended append\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The amino acid sequence is stored in a FASTA file and should be parsed unto our structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Parses fasta file line by line. \n",
    "    Only one line in memory at a time to suit for larger files\n",
    "    :param filename:  path to fasta file\n",
    "    :return: list of AASequences\n",
    "    \"\"\"\n",
    "    seq_list = []\n",
    "    curr_seq = None\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        line = f.readline()\n",
    "        while line:\n",
    "            line = line.strip()\n",
    "            # if sequence header, create new instance else continue parsing current one\n",
    "            if line.startswith('>'):\n",
    "                curr_seq = AASequence()         # create new QQSequence instance\n",
    "                seq_list.append(curr_seq)       # append reference to the AASequence object\n",
    "            else:\n",
    "                curr_seq.extend(line)\n",
    "            line = f.readline()\n",
    "    return seq_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concepts and definitions \n",
    "- genome : The complete set of genes or genetic material present in a cell or organism\n",
    "- Sequence Identity ratio : Ratio of residues which match exactly between the two sequences.\n",
    "- Sequence similarity : Two residues are similar if their substitution doesn't affect their function.\n",
    "- Sequence homology : Two sequences are homological\n",
    "- Protein domain : Domains, on the other hand, are regions of a protein that has a specific function and can (usually) function independently of the rest of the protein.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scores and substitution matrices\n",
    "Let us consider the 2 sequences:\n",
    " - **Seq1 : ATGACTGGA**\n",
    " - **Seq2 : ATATGA**\n",
    "\n",
    "There are multiple ways these sequences can be aligned:\n",
    "\n",
    "**ATGACTGGA** <br>\n",
    "**ATA - - TG- A**\n",
    "<br>or<br>\n",
    "**ATGACTGGA**<br>\n",
    "**AT- A- T- GA**\n",
    "\n",
    "*Which one is the best aligenemnt ? *\n",
    "\n",
    "### Scores based on identity\n",
    "\n",
    "<table>\n",
    "<caption>Score matrix for identity</caption>\n",
    "<tr>\n",
    "    <th>-</th>\n",
    "    <th>A</th>\n",
    "    <th>T</th>\n",
    "    <th>C</th>\n",
    "    <th>G</th>\n",
    "</tr>\n",
    "\n",
    "\n",
    "<tr>\n",
    "    <th>A</th>\n",
    "    <th>1</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>T</th>\n",
    "    <th>0</th>\n",
    "    <th>1</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>C</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>1</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>G</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>1</th>\n",
    "</tr>\n",
    "</table>\n",
    "\n",
    "An identity score matrix is computed according to the following equation:\n",
    "\n",
    "\\begin{equation}\n",
    "    identity\\_mat(i, j)= \n",
    "    \\begin{cases}\n",
    "        1, & \\text{if } i=j\\\\\n",
    "        0, & \\text{otherwise}\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "According to the identity matrix, the second alignement is better because it has a higher score.<br>\n",
    "Alignement 1 score:  $1+1+0+0+0+1+1+0+1=5$<br>\n",
    "Alignement 2 score: $1+1+0+1+0+1+0+1+1=6$\n",
    "\n",
    "\n",
    "\n",
    "### Scores based on similarity\n",
    "The previous method only favorises the conservation of amino acids, it does not take in consideration characteristics introduced by the evolution, like:\n",
    " - Each amino acid could be replaced by another one with the same biochemical characteristics *(e.g hydrophilic/hydrophobic molecules)*\n",
    " - Some amino acids are essential to the protein's stability, thus their substitution shan't be accepted by the evolution.\n",
    "\n",
    " \n",
    "it is, therefore, more appropriate to calculate scores based on **similarity**.\n",
    "Some mutations should be given a higher score than others, **rewarding** or **penalizing** their existence based on the probability of their occurrence.\n",
    "\n",
    "The probability of mutual substitution is influenced by numerous factors, it is impossible to theoretically calculate a substitution matrix. We rather derive such matrices by evaluating actual substitution rates observed on real protein sequence alignments.\n",
    "\n",
    "##### PAM matrices\n",
    "<img src=\"img/pam120.png\">\n",
    "\n",
    "The PAM matrices (Point Accepted Mutation) are computed based on the hypothesis that each amino acid in a position *p* takes place independently of previous mutations that happened at the same position *p*. \n",
    "This hypothesis qualifies the model to be defined as a *Markov process*.\n",
    "\n",
    "<u>Markovian process:</u>\n",
    "*Let $p$ the probability of occurrence of a Markovian event $e$, and $P[x]$ the probability that e occurs $x$ times.*\n",
    "Then $ P[x] = p^x $\n",
    "\n",
    "The PAM matrix $pam_1$ for the first generation can be computed based on observation of the first changes that occur as proteins diverge from a common ancestor.<br>\n",
    "PAM matrices that may be used to compare more distantly related proteins are derived according to the markovian rules by exponentiated the $pam_1$ matrix:<br>\n",
    "*Let $pam_x$ the matrice on x^{th} generation, then $pam_x=pam_1^x$*\n",
    "\n",
    "\n",
    "\n",
    "##### BLOSUM matrices\n",
    "The BLOSUM matrices (BLOcks SUbstitution Matrix).<br>\n",
    "<img src=\"img/blosum62.png\">\n",
    "\n",
    "With BLOSUM matrices, a positive score implies that substitution is **more likely** than any random substitution.\n",
    "A negative score implies that substitution is **less likely** than any random substitution.<br>\n",
    "Unlike PAM matrices, BLOSUM matrices are derived using local multiple alignments rather than global alignments.\n",
    "Local alignments observed are grouped together if they exceed an indicated limit for identity ratio.\n",
    "Several BLOSUM matrices can be produced by varying this limit, for instance, the BLOSUM62 matrix was derived using a 62% identity threshold.\n",
    "\n",
    "\n",
    "\n",
    "***Which substitution matrix to choose?***\n",
    "\n",
    "The PAM number indicates the divergence between the sequence, whereas the BLOSUM's number indicates the similarity ratio.\n",
    "Therefore it is generally more suitable to use PAM matrices for already known closely related protein sequences. Whereas BLOSUMs are more convenient to score alignments between evolutionarily divergent protein sequence because the matrix values the concept of functional similarity of amino acids.\n",
    "Thus PAM in our project we prefer using PAM matrices for global alignments and BLOSUM for local alignments.\n",
    "\n",
    "\n",
    "Let us define  structures representing the matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Matrix:\n",
    "    \"\"\" General matrix representation\"\"\"\n",
    "    def __init__(self, nrows, ncols, init_val=None):\n",
    "        self.nrows = nrows\n",
    "        self.ncols = ncols\n",
    "        self.init_val = init_val if not isinstance(init_val, MutableSequence) else deepcopy(init_val)\n",
    "        self.mat = None\n",
    "        self.reset()        # fill matrix in self.mat\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        return self.mat[idx]\n",
    "\n",
    "    def __setitem__(self, idx, newrow):\n",
    "        assert len(newrow) == self.nrows\n",
    "        self.mat[idx] = newrow\n",
    "\n",
    "    def __str__(self):\n",
    "        res = ''\n",
    "        for col in self.mat:\n",
    "            res += ' '.join([repr(x) for x in col]) + '\\n'\n",
    "        return res\n",
    "\n",
    "    def __repr__(self):\n",
    "        res = 'Matrix({})'\n",
    "        content = ', '.join([repr(col) for col in self.mat])\n",
    "        return res.format(content)\n",
    "\n",
    "    def reset(self):\n",
    "        self.mat = [[None for j in range(self.ncols)] for i in range(self.nrows)]\n",
    "        for i in range(self.nrows):\n",
    "            for j in range(self.ncols):\n",
    "                if isinstance(self.init_val, MutableSequence):\n",
    "                    self.mat[i][j] = deepcopy(self.init_val)\n",
    "                else:\n",
    "                    self.mat[i][j] = self.init_val\n",
    "\n",
    "\n",
    "    def fillrow(self, i, val):\n",
    "        \"\"\" fills row i with val \"\"\"\n",
    "        for j in range(self.ncols):\n",
    "            self.mat[i][j] = val\n",
    "\n",
    "    def fillcol(self, j, val):\n",
    "        \"\"\" fills col i with val \"\"\"\n",
    "        for i in range(self.nrows):\n",
    "            self.mat[i][j] = val\n",
    "\n",
    "    def getmax(self):\n",
    "        \"\"\" returns value, row, col of maximum in matirx\"\"\"\n",
    "        maxval = 0, 0, 0  # value, i, j\n",
    "        for i in range(self.nrows):\n",
    "            for j in range(self.ncols):\n",
    "                if self.mat[i][j] > maxval[0]:\n",
    "                    maxval = self.mat[i][j], i, j\n",
    "        return maxval\n",
    "\n",
    "\n",
    "class SubstitutionMatrix(Matrix):\n",
    "    __valid_AA = 'ABCDEFGHIKLMNPQRSTVWXYZ'\n",
    "\n",
    "    def __init__(self):\n",
    "        self.header_row = dict()\n",
    "        self.header_col = dict()\n",
    "        self.__size = len(self.__valid_AA) + 1  # +1 for the *\n",
    "        super().__init__(self.__size, self.__size, init_val=None)\n",
    "\n",
    "    def __getitem__(self, item):\n",
    "        if isinstance(item, int):  # standard matrix access with row and col number\n",
    "            return super().__getitem__(item)\n",
    "        elif isinstance(item, str):  # access with symbols submat['A']['R']\n",
    "            i = self.header_row[item]\n",
    "            return dict(zip(self.header_col, self[i]))\n",
    "        else:\n",
    "            raise TypeError\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The substitution matrices data is stored in text files, and should be parsed unto our structures as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_matrix_file(filename):\n",
    "    \"\"\"\n",
    "    Parses file to SubstitutionMatrix object\n",
    "    :param filename: path to the txt file\n",
    "    :return: SubstitutionMatrix object\n",
    "    \"\"\"\n",
    "    submat = SubstitutionMatrix()\n",
    "    n_line = -1\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        line = f.readline()\n",
    "        while line:\n",
    "            line = line.strip().split()\n",
    "            if line and line[0] != '#':\n",
    "                if n_line == -1:        # header line\n",
    "                    # dict(line[i]:i for i in range(len(line))\n",
    "                    submat.header_row = dict(zip(line, range(len(line))))       \n",
    "                else:\n",
    "                    # set in dictionary key:value of column symbol and number ex. 'R':1\n",
    "                    submat.header_col[line[0]] = n_line     \n",
    "                    for i in range(len(line[1:])):\n",
    "                        submat[n_line][i] = int(line[i+1])\n",
    "                n_line += 1\n",
    "            line = f.readline()\n",
    "    return submat\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence Alignement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "### Global alignement\n",
    "\n",
    "Global alignment is an alignment technique used to compare the sequences with apparent common ancestry.<br>\n",
    "It attempts to align every residue in every sequence and is mostly employed with sequences of roughly equal length.\n",
    "\n",
    "In our project we implement the Needleman-Wunsch algorithm, which is based on dynamic programming techniques, to produce the best $k$ global alignments of 2 given sequences.\n",
    "Let us consider 2 sequences $s1$ and $s2$ with length $m$ and $n$ respectively, on which we'll be applying the sequence alignement.\n",
    "\n",
    "#### Computing\n",
    "\n",
    "The Needleman-Wunsch algorithm attempts to find the alignment with the greater similarity.<br>\n",
    "In order to do so, we compare in a score matrix each amino acid of one sequence against all others of the second sequence, and an empty cell representing a gap.\n",
    "\n",
    "Therefore we use a *score matrix* of size $(n+1, m+1)$ which is a cartesian product of the 2 sequences, and an empty cell for indels.\n",
    "\n",
    "The structure is defined as follow:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ScoreMatrix(Matrix):\n",
    "    def __init__(self, seq1, seq2):\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        super().__init__(ncols=len(seq1) + 1, nrows=len(seq2) + 1, init_val=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The comparison in the score matrix $S$ is based on the following mathematical score functions.\n",
    "\\begin{equation}\n",
    "    S(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j-1)+t(i,j) \\\\\n",
    "        [S(I-N_{gap1}, j)+g(n_{gap1})]_{1 \\le ngap1 \\le i}\\\\\n",
    "        [S(i,j-n_{gap2})+g(n_{gap2})]_{1 \\le ngap2 \\le j}\\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "Where $t(i,j)$ is the score in the substitution matrix:\n",
    "\\begin{equation}\n",
    "t(i,j)=MATSUB[s1[i], s2[j]]\n",
    "\\end{equation}\n",
    "\n",
    "And $g$ is the affine penalty for gaps, defined as the following:\n",
    "\\begin{equation}\n",
    "g(n_{gap}) = - I  -(n_{gap}-1) \\cdot E \\text{ where } I \\in [7,15] \\text{ and } E \\in [0.5,2]\\\\\n",
    "\\end{equation}\n",
    "\n",
    "\n",
    "<img src=\"img/global_mat_rec.png\">\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gaps are scored in a much more heuristic way than substitutions. In contrast to a linear penalty in which all gaps penalize the score equally, the affine penalty technique (reasonably) penalizes the opening of a gap more heavily than its extension, because once a gap is introduced, it is easier (more probable) for it to be extended.<br>\n",
    "The affine penalty function makes use of multiple variables :\n",
    "- $n_{gap}$  : The number of the gap in the mutation\n",
    "- $I$ : Penalty of the first gap\n",
    "- $E$ : Penalty of an extension of a gap, depending on how much it got extended.\n",
    "\n",
    "The values of gap parameters need to be carefully chosen. A poorly chosen parameter can significantly degrade the performance of the algorithm."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The execution time of filling the $S$ matrix is $O(mn*mn=mn^2)$. The algorithm can be optimized to $O(mn)$ by splitting the recursion into 3 separate functions and using dynamic programming with each one of them.\n",
    "Each of these functions will save their results to their separate score matrix.\n",
    "The V and W matrices are used to hold values compared to gaps, while the S matrix is for substitutions/matches.\n",
    "\n",
    "\\begin{equation}\n",
    "    V(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j)-1 \\\\\n",
    "        V(i-1,j)-E \\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "\\begin{equation}\n",
    "    W(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i,j-1)-1 \\\\\n",
    "        W(i,j-1)-E \\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "\\begin{equation}\n",
    "    S(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j-1)+t(i,j) \\\\\n",
    "        V(i,j) \\\\\n",
    "        W(i,j)\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "\n",
    "\n",
    "<img src=\"img/vws.png\" >\n",
    "\n",
    "##### Matrices initialization\n",
    "The matrices need to be initialized before they can be filled during the computation.\n",
    "All the matrices hold `None` as a default value, representing the fact that no processing on the cell has already been made.\n",
    "\n",
    "The S matrix' first column $(i=*,j=0)$ and row $(i=0,j=*)$ are filled with $g(i)$ and $g(j)$ respectively because they represent a gap insertion (and extension).\n",
    "\n",
    "The W matrix' first col $(i=*,j=0)$ and V matrix' first row $(i=0,j=*)$ are filled with `float('-inf')` to force the S function for cells in the second column and row $(i=*,j=1)$ and $(i=1,j=*)$ of the S matrix into not selecting $V(i,j)$ or $W(i,j)$ as a result.\n",
    "The reason is that selecting one of them represents a gap which would be inconvenient in the first position of the alignment. We require that the alignment does not start with a (or many) gaps, therefore we force the $S$ function into calculating the value with $S(i-1,j-1)+t(i,j)$.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Ranking solutions\n",
    "The Needleman-Wunsch algorithm produces multiple alignments. We chose to rank the solutions based on the identity degree.\n",
    "Identity describes the ratio to which amino acids match at a same position.<br>\n",
    "One might note that 2 completely unrelated sequences might not have an identity ratio of 0 because the probability of having 2 molecules that match at a same position is not 0, thus making the measures based on identity unreliable.\n",
    "However, having a relatively high identity with long sequences is considerably less probable than with shorter sequences (*Binomial law for random variables*).\n",
    "\n",
    "Similarity is another concept to compare 2 sequences. Similarity represents the ratio of functional similarity of 2 sequences, based on the substitution matrix used.\n",
    "\n",
    "\n",
    "The identity and similarity functions are defined as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_identity(s1, s2):\n",
    "    assert len(s1) == len(s2)\n",
    "    c = 0\n",
    "    for c1, c2 in zip(s1, s2):\n",
    "        if c1 == c2:\n",
    "            c += 1\n",
    "    return c/len(s1)\n",
    "\n",
    "\n",
    "def get_similarity(s1, s2, submat):\n",
    "    assert len(s1) == len(s2)\n",
    "    c = 0\n",
    "    for c1,c2 in zip(s1, s2):\n",
    "        if c1 != '-' and c2 != '-':\n",
    "            if submat[c1][c2] >= 0:\n",
    "                c += 1\n",
    "    return c/len(s1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting solutions\n",
    "Each cell in the score matrix (except the cell at position (0,0)) is computed by selecting the maximum between its top, left or top-left cell. In a way, each cell can originate from one of these neighboring cells, or many, in case they produce the same equal maximum value.\n",
    "\n",
    "We can reconstruct solutions by following all possible reverse paths starting from the last cell in the score matrix $S$ (position (n,m)) leading to the first cell (position (0,0)), from which all cells are descendants.\n",
    "\n",
    "<img src=\"img/tracepath_global.png\">\n",
    "It is therefore beneficial to log the potential origins of each cell during the computation of the matrix $S$ to be able to construct solutions in an optimized way.\n",
    "We can thus use a helper matrix to store the possible origins of each cell. Such a matrix is defined in the following as *PathMatrix*.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PathMatrix(Matrix):\n",
    "    DIAG, LEFT, TOP = 0, 1, 2\n",
    "\n",
    "    def __init__(self, seq1, seq2):\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.init_val = [False, False, False]\n",
    "        super().__init__(ncols=len(seq1)+1, nrows=len(seq2)+1, init_val=self.init_val)\n",
    "\n",
    "    def isdiag(self, i, j):\n",
    "        return self.mat[i][j][self.DIAG]\n",
    "\n",
    "    def isleft(self, i, j):\n",
    "        return self.mat[i][j][self.LEFT]\n",
    "\n",
    "    def istop(self, i, j):\n",
    "        return self.mat[i][j][self.TOP]\n",
    "\n",
    "    def isnone(self, i, j):\n",
    "        return not any(self.mat[i][j])      # [False, False, False]\n",
    "\n",
    "    def setdiag(self, i, j):\n",
    "        self.mat[i][j][self.DIAG] = True\n",
    "\n",
    "    def setleft(self, i, j):\n",
    "        self.mat[i][j][self.LEFT] = True\n",
    "\n",
    "    def settop(self, i, j):\n",
    "        self.mat[i][j][self.TOP] = True\n",
    "\n",
    "    def erase(self, i, j):\n",
    "        \"\"\" reset cell to default [false, false, false] \"\"\"\n",
    "        self.mat[i][j] = deepcopy([False, False, False])\n",
    "\n",
    "    def isdivergent(self, i, j):\n",
    "        \"\"\" checks if multiple origins possible to the cell\"\"\"\n",
    "        return self.mat[i][j].count(True) > 1\n",
    "\n",
    "    def fillrow(self, i, diag, left, top):\n",
    "        for j in range(self.ncols):\n",
    "            # diag, left and top are booleans\n",
    "            x = [None, None, None]\n",
    "            x[self.DIAG] = diag\n",
    "            x[self.TOP] = top\n",
    "            x[self.LEFT] = left\n",
    "            self.mat[i][j] = deepcopy(x)\n",
    "\n",
    "    def fillcol(self, j, diag, left, top):\n",
    "        for i in range(self.nrows):\n",
    "            # diag, left and top are booleans\n",
    "            x = [None, None, None]\n",
    "            x[self.DIAG] = diag\n",
    "            x[self.TOP] = top\n",
    "            x[self.LEFT] = left\n",
    "            self.mat[i][j] = deepcopy(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Path matrix initialization\n",
    "The path matrix is initialized with default value `[False, False, False]`, representing the fact that not any path (diag, top, left) has already been set.<br>\n",
    "The cells in the first row are set to `path=left` because they are linearly calculated from the origin (position 0,0 of the matrix) ongoing towards the right. Each cell originates from its left neighbor.<br>\n",
    "The cells in the first column are set to `path=top` because they are linearly calculated from the origin (position 0,0 of the matrix) ongoing towards the bottom. Each cell originates from its top neighbor.\n",
    "\n",
    "<table>\n",
    "<caption>Initial path matrix</caption>\n",
    "<tr>\n",
    "    <th>-</th>\n",
    "    <th>A</th>\n",
    "    <th>T</th>\n",
    "    <th>C</th>\n",
    "    <th>G</th>\n",
    "</tr>\n",
    "\n",
    "\n",
    "<tr>\n",
    "    <th>A</th>\n",
    "    <th>0</th>\n",
    "    <th>Left</th>\n",
    "    <th>Left</th>\n",
    "    <th>Left</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>T</th>\n",
    "    <th>Top</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>C</th>\n",
    "    <th>Top</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "\n",
    "<tr>\n",
    "    <th>G</th>\n",
    "    <th>Top</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "    <th>0</th>\n",
    "</tr>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Reconstruction Tree\n",
    "\n",
    "Our reconstruction algorithm is inspired by compression algorithms, it uses a tree structure (defined below) to store all the possible solutions efficiently."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Tree:\n",
    "    def __init__(self, ):\n",
    "        self.seq1 = ''\n",
    "        self.seq2 = ''\n",
    "        self.parent = None\n",
    "        self.children = []\n",
    "\n",
    "    def add_child(self, subtree):\n",
    "        self.children.append(subtree)\n",
    "        subtree.parent = self\n",
    "\n",
    "    def is_leaf(self):\n",
    "        return len(self.children) == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In a simplified way, the algorithm follows the path up in the score matrix while reconstructing the string storing it in a tree node, until it reaches a cell where multiple paths are possible. On that occasion, 2 children nodes are created and a recursive call is done for each one of them, reconstructing the divergent solutions in their respective nodes.\n",
    "\n",
    "For instance, let us consider 2 sequences **ATGACTGGA** and **ATATGA** for which we suppose exists multiple possible alignments:\n",
    "\n",
    "**ATGACTGGA** <br>\n",
    "**ATA - - TG- A**\n",
    "<br>and<br>\n",
    "**ATGACTGGA** <br>\n",
    "**ATA - - TGA-**\n",
    "<br>and<br>\n",
    "**ATGACTGGA**<br>\n",
    "**AT- A- T- GA**\n",
    "\n",
    "The reconstruction algorithm produces the following tree:\n",
    "<img src=\"img/tree.png\">\n",
    "Then, to enumerate all possible solutions, we compute all possible paths from root to leaf in the tree, concatenating the content nodes included in the path. Enumerating tree paths is implemented in the recursive function `find_paths`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_paths(tree):\n",
    "    \"\"\" returns a list of sequences concatenated from each possible path root->leaf\"\"\"\n",
    "    if tree.is_leaf():\n",
    "        return [(tree.seq1, tree.seq2)]\n",
    "    pathlist = []\n",
    "    for child in tree.children:\n",
    "        paths = find_paths(child)\n",
    "        for p in paths:\n",
    "            newpath = (tree.seq1 + p[0], tree.seq2 + p[1])\n",
    "            pathlist.append(newpath)\n",
    "    return pathlist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Proof of Concept\n",
    "\n",
    "The complete Needleman-Wunsch implementation is defined in a `NeedlemanWunsch` class.<br>\n",
    "It takes in parameter 2 sequences to align, the error parameters $I$ and $E$, the substitution matrix from which the algorithm operate and a parameter $k$ for specifying how many sequences should be returned.<br>\n",
    "The $k$ solution returned are sorted from best to worst.\n",
    "\n",
    "We chose to represent solutions in a `Solution` structure to facilitate usage and readability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_pretty_string(seq1, seq2, submat):\n",
    "    \"\"\"\n",
    "    Returns readable alignment string given 2 aligned sequences and the substitution matrix used\n",
    "    The format is the same as the one used in LALIGN\n",
    "    | molecule match\n",
    "    : mutation, positive score in substitution matrix\n",
    "    . mutation, negative score in substitution matrix\n",
    "    <space> mutation, indel    \n",
    "    \"\"\"\n",
    "    assert len(seq1) == len(seq2)\n",
    "    s1 = list(seq1)\n",
    "    s3 = [' ' for _ in range(len(seq1))]\n",
    "    s2 = list(seq2)\n",
    "\n",
    "    for i in range(len(seq1)):\n",
    "        if s1[i] == '-' or s2[i] == '-':\n",
    "            continue\n",
    "        elif s1[i] == s2[i]:\n",
    "            s3[i] = ':'\n",
    "        else:\n",
    "            if submat[seq1[i]][seq2[i]] >= 0:\n",
    "                s3[i] = '.'\n",
    "            else:\n",
    "                s3[i] = ' '\n",
    "\n",
    "    pretty_sol = ''.join(s1) + '\\n' + ''.join(s3) + '\\n' + ''.join(s2)\n",
    "    return pretty_sol\n",
    "    \n",
    "\n",
    "class Solution:\n",
    "    \"\"\" Structure representing one solution of aligning 2 sequences\"\"\"\n",
    "    def __init__(self, s1, s2, submat):\n",
    "        self.s1 = s1\n",
    "        self.s2 = s2\n",
    "        self.submat = submat\n",
    "        self.identity = get_identity(s1, s2)\n",
    "        self.similarity = get_similarity(s1, s2, submat)\n",
    "        self.prettystr = build_pretty_string(s1, s2, submat)\n",
    "\n",
    "    def __str__(self):\n",
    "        res = self.prettystr + '\\n'\n",
    "        res += 'Identity : %s %%\\n' % round(self.identity * 100, 2)\n",
    "        res += 'Similarity : %s %%\\n' % round(self.similarity * 100, 2)\n",
    "        res += 'Length : %d\\n' % len(self.s1)\n",
    "        return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class NeedlemanWunsch:\n",
    "    \"\"\"\n",
    "    NeedlemanWunsch implementation\n",
    "    [*] usage : \n",
    "        nw = NeedlemanWunsch(...)\n",
    "        nw.run()\n",
    "        nw.display()\n",
    "    \"\"\"\n",
    "    def __init__(self, seq1, seq2, I, E, submat, k):\n",
    "        # storing params\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.I = I\n",
    "        self.E = E\n",
    "        self.submat = submat\n",
    "        self.k = k\n",
    "        self.solutions = [] # list of Solution objects\n",
    "    \n",
    "        # initializing computation matrices\n",
    "        self.V_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.W_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.S_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.W_mat.fillcol(j=0, val=float('-inf'))\n",
    "        self.V_mat.fillrow(i=0, val=float('-inf'))\n",
    "        self.init_S_mat()\n",
    "        \n",
    "        # initializing path matrix\n",
    "        self.path_mat = PathMatrix(seq1=seq1, seq2=seq2)\n",
    "        self.path_mat.fillcol(0, diag=False, top=True, left=False)\n",
    "        self.path_mat.fillrow(0, diag=False, top=False, left=True)\n",
    "\n",
    "    def display(self):\n",
    "        \"\"\" display solutions \"\"\"\n",
    "        hr = '-' * 20    # horizontal line\n",
    "        print(hr, '[*] Solution(s)', hr, sep='\\n')\n",
    "        print('s1 : %s' % s1, 's2 : %s' % s2, sep='\\n')\n",
    "        print('# solutions : ', len(self.solutions))\n",
    "        print()\n",
    "        for sol in self.solutions:\n",
    "            print(sol)\n",
    "        print(hr)\n",
    "\n",
    "    def init_S_mat(self):\n",
    "        \"\"\"Initialise S matrix\"\"\"\n",
    "        self.S_mat[0][0] = 0\n",
    "        for col in range(1, self.S_mat.ncols):\n",
    "            self.S_mat[0][col] = self.affine_gap_penalty(col)\n",
    "        for row in range(1, self.S_mat.nrows):\n",
    "            self.S_mat[row][0] = self.affine_gap_penalty(row)\n",
    "\n",
    "    def affine_gap_penalty(self, n):\n",
    "        return -self.I - (n - 1) * self.E\n",
    "\n",
    "    def t(self, i, j):\n",
    "        # -1 because the score matric has an extra row and column\n",
    "        return self.submat[str(self.seq2[i - 1])][str(self.seq1[j - 1])]\n",
    "\n",
    "    def V(self, i, j):\n",
    "        if self.V_mat[i][j] is None:\n",
    "            self.V_mat[i][j] = max(self.S(i - 1, j) - self.I, self.V(i - 1, j) - self.E)\n",
    "        return self.V_mat[i][j]\n",
    "\n",
    "    def W(self, i, j):\n",
    "        if self.W_mat[i][j] is None:\n",
    "            self.W_mat[i][j] = max(self.S(i, j - 1) - self.I, self.W(i, j - 1) - self.E)\n",
    "        return self.W_mat[i][j]\n",
    "\n",
    "    def S(self, i, j):\n",
    "        if self.S_mat[i][j] is None:\n",
    "            diag_score = self.S(i - 1, j - 1) + self.t(i, j)\n",
    "            top_score = self.V(i, j)\n",
    "            left_score = self.W(i, j)\n",
    "            self.S_mat[i][j] = max(diag_score, top_score, left_score)\n",
    "\n",
    "            # log path\n",
    "            # we consider multiple possibilities\n",
    "            if self.S_mat[i][j] == diag_score:\n",
    "                self.path_mat.setdiag(i, j)\n",
    "            if self.S_mat[i][j] == top_score:\n",
    "                self.path_mat.settop(i, j)\n",
    "            if self.S_mat[i][j] == left_score:\n",
    "                self.path_mat.setleft(i, j)\n",
    "\n",
    "        return self.S_mat[i][j]\n",
    "\n",
    "    def reconstruct(self, i, j, tree):\n",
    "        \"\"\" reconstructs tree of possible paths given an initial matrix position (i,j)\"\"\"\n",
    "        if not (i == 0 and j == 0):  \n",
    "            if not self.path_mat.isdivergent(i, j): # one possibility : continue on current node\n",
    "                if self.path_mat.isdiag(i, j):\n",
    "                    tree.seq1 += str(self.seq1[j - 1])\n",
    "                    tree.seq2 += str(self.seq2[i - 1])\n",
    "                    self.reconstruct(i - 1, j - 1, tree=tree)\n",
    "\n",
    "                elif self.path_mat.isleft(i, j):\n",
    "                    tree.seq1 += str(self.seq1[j - 1])\n",
    "                    tree.seq2 += '-'\n",
    "                    self.reconstruct(i, j - 1, tree=tree)\n",
    "\n",
    "                elif self.path_mat.istop(i, j):\n",
    "                    tree.seq1 += '-'\n",
    "                    tree.seq2 += str(self.seq2[i - 1])\n",
    "                    self.reconstruct(i - 1, j, tree=tree)\n",
    "                else:\n",
    "                    raise Exception('Logic error')\n",
    "\n",
    "            else:  # multiple possibilities: one forked node for each\n",
    "                if self.path_mat.isdiag(i, j):\n",
    "                    subtree = Tree()\n",
    "                    tree.add_child(subtree)\n",
    "                    subtree.seq1 += str(self.seq1[j - 1])\n",
    "                    subtree.seq2 += str(self.seq2[i - 1])\n",
    "                    self.reconstruct(i - 1, j - 1, tree=subtree)\n",
    "\n",
    "                if self.path_mat.isleft(i, j):\n",
    "                    subtree = Tree()\n",
    "                    tree.add_child(subtree)\n",
    "                    subtree.seq1 += str(self.seq1[j - 1])\n",
    "                    subtree.seq2 += '-'\n",
    "                    self.reconstruct(i, j - 1, tree=subtree)\n",
    "\n",
    "                if self.path_mat.istop(i, j):\n",
    "                    subtree = Tree()\n",
    "                    tree.add_child(subtree)\n",
    "                    subtree.seq1 += '-'\n",
    "                    subtree.seq2 += str(self.seq2[i - 1])\n",
    "                    self.reconstruct(i - 1, j, tree=subtree)\n",
    "        return tree\n",
    "\n",
    "    def run(self):\n",
    "        # compute inital S matrix\n",
    "        for row in range(1, self.S_mat.nrows):\n",
    "            for col in range(1, self.S_mat.ncols):\n",
    "                self.S_mat[row][col] = self.S(row, col)\n",
    "\n",
    "        tree = Tree()\n",
    "        self.reconstruct(i=self.S_mat.nrows - 1, j=self.S_mat.ncols - 1, tree=tree)\n",
    "        for s1, s2 in find_paths(tree):\n",
    "            self.solutions.append(Solution(s1=s1[::-1], s2=s2[::-1], submat=self.submat))\n",
    "        self.solutions.sort(key=lambda sol: sol.identity, reverse=True)\n",
    "        self.solutions = self.solutions[:self.k]\n",
    "        self.display()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Local alignment\n",
    "\n",
    "Unlike the Needleman-Wunsch algorithm which looks at the integrality of the sequence, the Smith-Waterman algorithm correlates segments of all possible sizes and selects the segment that yields the best similarity level. This technique is called *local alignment*.<br>\n",
    "The algorithm is used to identify regions of similarity that may be a consequence of functional, structural, or evolutionary relationships between the sequences.\n",
    "Therefore, in local alignments techniques, the score functions are altered to only allow positive scores to be considered in the score matrix.\n",
    "The reason is that negative scores produce alignments where the high similarity regions are *sparsely* spread across the alignment. In this case, we prefer to find regions of the protein that are similar in a \"condensed\" way. Therefore, negative values are set to 0 and not considered.\n",
    "\n",
    "\n",
    "\n",
    "#### Computation\n",
    "The score functions for local alignments can be defined as follows:\n",
    "\n",
    "\\begin{equation}\n",
    "    V(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j)-1 \\\\\n",
    "        V(i-1,j)-E \\\\\n",
    "        \\color{red}0 \\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "\\begin{equation}\n",
    "    W(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i,j-1)-1 \\\\\n",
    "        W(i,j-1)-E \\\\\n",
    "        \\color{red}0 \\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "\\begin{equation}\n",
    "    S(i, j)= max\n",
    "    \\begin{cases}\n",
    "        S(i-1,j-1)+t(i,j) \\\\\n",
    "        V(i,j) \\\\\n",
    "        W(i,j) \\\\\n",
    "        \\color{red}0 \\\\\n",
    "    \\end{cases}\n",
    "\\end{equation}\n",
    "\n",
    "##### Matrices initialization\n",
    "The score matrices V, W and S are initialized according to the same reasoning as with the global alignment.\n",
    "We note that the first row and column of the S matrix are filled with 0 because their initial theoretical value (calculated by the gap function) is always negative.\n",
    "\n",
    "\n",
    "\n",
    "#### Extracting solutions\n",
    "When computing matches (retracing reverse paths), the tracing stops when it reaches a step where no positive possibility is available. It then \"marks\" the cells of the current path, resetting them to 0 to not be considered in future solutions.<br>\n",
    "Considering the changes applied in the matrix, the algorithm then updates the cells concerned by the modifications.\n",
    "\n",
    "<img src=\"img/tracepath_zero_local.png\">\n",
    "\n",
    "\n",
    "It is important to note that in contrast with global alignment, we only favor <u>*one*</u> possible path when tracing back the origin of the alignment. Facing a case where a cell could have multiple origins, the path to be privileged is selected based on the following ranking:\n",
    "1. diagonal\n",
    "2. left\n",
    "3. top\n",
    "\n",
    "We give higher priority to a diagonal origin because it doesn't involve any gap insertion, keeping the similarity densely spread across the alignment.\n",
    "The left and top are theoretically of equal privilege and could as well be swapped, but have be ranked as such due to algorithmic requirements.\n",
    "Extracting solutions in the Smith-Waterman algorithm doesn't require the construction of a tree.\n",
    "\n",
    "\n",
    "In our implementation, generating a solution is done in 2 functions, `tracepath` and `next_sequence`.\n",
    "The tracepath is given a starting position in the score matrix, which is the maximal value present in the matrix (representing the best score produced).\n",
    "    The function then traces back the origin of the subsequence by checking the pathmatrix `path_mat` until a 0 is reached.\n",
    "The alignment strings and a list of positions of the path in (row, column) format are returned.\n",
    "\n",
    "\n",
    "*Pseudo code for tracepath:*\n",
    "```\n",
    "def tracepath(S_mat, path_mat, i, j):\n",
    "    s1, s2 = '', ''\n",
    "    path = []\n",
    "    while not S_mat[i][j] == 0:\n",
    "        path.append((i,j))\n",
    "        if path_mat.isdiag(i, j):\n",
    "            append char of row in s1\n",
    "            append char of col in s2\n",
    "            go up to diag top left\n",
    "        elif path_mat.isleft(i, j):\n",
    "            append char of row in s1\n",
    "            append '-' of col in s2\n",
    "            go left\n",
    "        elif path_mat.istop(i, j):\n",
    "            append  '-' in s1\n",
    "            append char of col in s2\n",
    "            go up\n",
    "    return s1, s2, path\n",
    " ```\n",
    " \n",
    " \n",
    " The `tracepath` function is a helper function used in the `next_sequence` function.\n",
    " `next_sequence` is responsible of generating a solution by first finding the best score in the score matrix (the maximum), calling tracepath on it, then marking the path positions in a separate matrix `blacklist` to not be considered in future alignments.<br>\n",
    "\n",
    "Let `init_pos=path[-1]` be the position of the cell encountered immediately before stopping at the 0.<br>\n",
    "The submatrix defined by the upper-left corner `init_pos` and lower-right corner `(n, m)` is recalculated entirely to update the cells affected by the modifications.\n",
    "\n",
    " *Pseudo code of next_sequence*:\n",
    " \n",
    " ```\n",
    " def next_sequence(S_mat):\n",
    "        \"\"\" Sequence generator \"\"\"\n",
    "        # get max and calculate path\n",
    "        value, maxi, maxj = S_mat.getmax()\n",
    "        s1, s2, path = tracepath(i=maxi, j=maxj)\n",
    "        # reverse the strings because the path is calculated in reverse order\n",
    "        s1, s2 = s1[::-1], s2[::-1]  \n",
    "\n",
    "        # mark and set to 0 traversed positions\n",
    "        for i, j in path:\n",
    "            # set value to 0 in all matrices and set True in blacklist\n",
    "            setzero(i, j)  \n",
    "\n",
    "        #erase submatrix starting from last position before reaching the 0 to absolute bottom right position\n",
    "        init_i, init_j = path[-1] # before last position\n",
    "        \n",
    "        # recalculate submatrix matrix\n",
    "        for i in range(init_i, self.S_mat.nrows):\n",
    "            for j in range(init_j, self.S_mat.ncols):\n",
    "                if not blacklist[i][j]:\n",
    "                    S_mat[i][j] = S(i, j)\n",
    "\n",
    "        return s1, s2\n",
    "```\n",
    "\n",
    "#### Proof of Concept\n",
    "The full Waterman-Smith algorithm is implemented as follows.\n",
    "In addition to the sequences, it takes in argument the error parameters, the substitution matrix to be used and a parameter $l$ which represents the number of solutions to be returned."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class SmithWaterman:\n",
    "    def __init__(self, seq1, seq2, I, E, submat, l):\n",
    "        # storing parameters\n",
    "        self.seq1 = seq1\n",
    "        self.seq2 = seq2\n",
    "        self.I = I\n",
    "        self.E = E\n",
    "        self.submat = submat\n",
    "        self.l = l\n",
    "        self.solutions = [] # list of Solution objects\n",
    "        \n",
    "        # initializing computation matrices\n",
    "        self.V_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.W_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.S_mat = ScoreMatrix(seq1, seq2)\n",
    "        self.W_mat.fillcol(j=0, val=float('-inf'))\n",
    "        self.V_mat.fillrow(i=0, val=float('-inf'))\n",
    "        self.S_mat.fillrow(i=0, val=0)\n",
    "        self.S_mat.fillcol(j=0, val=0)\n",
    "        \n",
    "        # initializing path matrix and blacklist\n",
    "        self.path_mat = PathMatrix(seq1, seq2)\n",
    "        self.blacklist = Matrix(ncols=len(seq1)+1, nrows=len(seq2)+1, init_val=False)        \n",
    "\n",
    "    def display(self):\n",
    "        \"\"\" display solutions \"\"\"\n",
    "        hr = '-' * 20\n",
    "        print(hr, '[*] Solution(s)', hr, sep='\\n')\n",
    "        print('s1 : %s' % self.seq1, 's2 : %s' % self.seq2, 'I=%d' % self.I, 'E=%d' % self.E, sep='\\n')\n",
    "        print('# solutions : ', len(self.solutions))\n",
    "        print()\n",
    "        for sol in self.solutions:\n",
    "            print(sol, end='\\n\\n')\n",
    "        print(hr)\n",
    "\n",
    "    def t(self, i, j):\n",
    "        # -1 because the score matrix has an extra row and column\n",
    "        return self.submat[str(self.seq2[i - 1])][str(self.seq1[j - 1])]\n",
    "\n",
    "    def V(self, i, j):\n",
    "        if self.V_mat[i][j] is None:\n",
    "            self.V_mat[i][j] = max(0, self.S(i - 1, j) - self.I, self.V(i - 1, j) - self.E)\n",
    "        return self.V_mat[i][j]\n",
    "\n",
    "    def W(self, i, j):\n",
    "        if self.W_mat[i][j] is None:\n",
    "            self.W_mat[i][j] = max(0, self.S(i, j - 1) - self.I, self.W(i, j - 1) - self.E)\n",
    "        return self.W_mat[i][j]\n",
    "\n",
    "    def S(self, i, j):\n",
    "        if self.S_mat[i][j] is None:\n",
    "            diag_score = self.S(i - 1, j - 1) + self.t(i, j)\n",
    "            top_score = self.V(i, j)\n",
    "            left_score = self.W(i, j)\n",
    "            self.S_mat[i][j] = max(0, diag_score, top_score, left_score)\n",
    "\n",
    "            # log path\n",
    "            # we dont take path divergences in consideration\n",
    "            # diag has priority\n",
    "            if self.S_mat[i][j] == diag_score:\n",
    "                self.path_mat.setdiag(i, j)\n",
    "            elif self.S_mat[i][j] == top_score:\n",
    "                self.path_mat.settop(i, j)\n",
    "            elif self.S_mat[i][j] == left_score:\n",
    "                self.path_mat.setleft(i, j)\n",
    "        return self.S_mat[i][j]\n",
    "\n",
    "    def setzero(self, i, j):\n",
    "        \"\"\" mark cell as already treated, and set values to 0\"\"\"\n",
    "        self.blacklist[i][j] = True\n",
    "        self.V_mat[i][j] = 0\n",
    "        self.W_mat[i][j] = 0\n",
    "        self.S_mat[i][j] = 0\n",
    "        self.path_mat.erase(i, j)\n",
    "\n",
    "    def tracepath(self, i, j):\n",
    "        \"\"\" Given initial position (i,j) find sequence alignement and the path taken\n",
    "        according to the SmithWaterman algorithm\"\"\"\n",
    "        s1, s2 = '', ''\n",
    "        path = []\n",
    "\n",
    "        while not self.S_mat[i][j] == 0:\n",
    "            path.append((i,j))\n",
    "            if self.path_mat.isdiag(i, j):\n",
    "                s1 += str(self.seq1[j - 1])\n",
    "                s2 += str(self.seq2[i - 1])\n",
    "                i = i-1\n",
    "                j = j-1\n",
    "\n",
    "            elif self.path_mat.isleft(i, j):\n",
    "                s1 += str(self.seq1[j - 1])\n",
    "                s2 += '-'\n",
    "                i = i\n",
    "                j = j-1\n",
    "\n",
    "            elif self.path_mat.istop(i, j):\n",
    "                s1 += '-'\n",
    "                s2 += str(self.seq2[i - 1])\n",
    "                i = i-1\n",
    "                j = j\n",
    "            else:\n",
    "                raise Exception('Logic error')\n",
    "        return s1, s2, path\n",
    "\n",
    "    def next_sequence(self):\n",
    "        \"\"\" Sequence generator \"\"\"\n",
    "        # get max and calculate path\n",
    "        value, maxi, maxj = self.S_mat.getmax()\n",
    "        s1, s2, path = self.tracepath(i=maxi, j=maxj)\n",
    "        s1, s2 = s1[::-1], s2[::-1]\n",
    "\n",
    "        # mark and set to 0\n",
    "        for i, j in path:\n",
    "            self.setzero(i, j)\n",
    "\n",
    "        # erase submatrix starting from last position before reaching the 0 to absolute bottom right position\n",
    "        init_i, init_j = path[-1]\n",
    "        for i in range(init_i, self.S_mat.nrows):\n",
    "            for j in range(init_j, self.S_mat.ncols):\n",
    "                if not self.blacklist[i][j]:\n",
    "                    self.S_mat[i][j] = None\n",
    "                    self.W_mat[i][j] = None\n",
    "                    self.V_mat[i][j] = None\n",
    "        # recalculate matrix\n",
    "        for i in range(init_i, self.S_mat.nrows):\n",
    "            for j in range(init_j, self.S_mat.ncols):\n",
    "                if not self.blacklist[i][j]:\n",
    "                    self.S_mat[i][j] = self.S(i, j)\n",
    "\n",
    "        return s1, s2\n",
    "\n",
    "    def run(self):\n",
    "        # fill the S score matrix\n",
    "        for row in range(1, self.S_mat.nrows):\n",
    "            for col in range(1, self.S_mat.ncols):\n",
    "                self.S_mat[row][col] = self.S(row, col)\n",
    "\n",
    "        # generate l sequences\n",
    "        for _ in range(self.l):\n",
    "            s1, s2 = self.next_sequence()\n",
    "            self.solutions.append(Solution(s1, s2, self.submat))\n",
    "\n",
    "        self.display()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results and discussion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We start by parsing the given FASTA files (`WW-sequence.fasta` and `protein-sequences.fasta`) and substitution matrices PAM60, PAM120, BLOSUM62, BLOSUM80.\n",
    "These are needed in our algorithms to produce alignments.<br>\n",
    "We will be testing our algorithms with different parameters and comparing the results with LALIGN, then try to make sense of the resuls with based on information available on UniProt."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parsing FASTA files\n",
    "seqlist_WW = parse_fasta_file('rsrc/WW-sequence.fasta')\n",
    "seqlist_protein = parse_fasta_file('rsrc/protein-sequences-o.fasta')\n",
    "\n",
    "# parsing substitution matrices\n",
    "pam60 = parse_matrix_file('rsrc/pam60.txt')\n",
    "pam120 = parse_matrix_file('rsrc/pam120.txt')\n",
    "blosum62 = parse_matrix_file('rsrc/blosum62.txt')\n",
    "blosum80 = parse_matrix_file('rsrc/blosum80.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### What is LALIGN\n",
    "LALIGN is an online platform that offers global and local alignment of sequences with customizable parameters and matrices.\n",
    "\n",
    "##### What is UniProt\n",
    "UniProt is an online comprehensive resource for protein sequence and genetical data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global alignment\n",
    "For each pair of sequences in the `WW-sequence.fasta` file (*15 sequences in total*), we compute up to $k=5$ solutions using the Needleman-Wunsch algorithm with varying parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 1\n",
    "# Substitution matrix : PAM120\n",
    "# I: 12\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: 18\n",
    "\n",
    "\n",
    "for s1, s2 in combinations(seqlist_WW, 2):\n",
    "    NeedlemanWunsch(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=12,\n",
    "        E=2,\n",
    "        submat=pam120,\n",
    "        k=5\n",
    "    ).run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 2\n",
    "# Substitution matrix : PAM120\n",
    "# I: 8\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: 26\n",
    "\n",
    "for s1, s2 in combinations(seqlist_WW, 2):\n",
    "    NeedlemanWunsch(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=8,\n",
    "        E=2,\n",
    "        submat=pam60,\n",
    "        k=5\n",
    "    ).run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 3\n",
    "# Substitution matrix : PAM60\n",
    "# I: 12\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: 21\n",
    "\n",
    "for s1, s2 in combinations(seqlist_WW, 2):\n",
    "    NeedlemanWunsch(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=12,\n",
    "        E=2,\n",
    "        submat=pam60,\n",
    "        k=5\n",
    "    ).run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 4\n",
    "# Substitution matrix : PAM60\n",
    "# I: 8\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: 30\n",
    "\n",
    "nw_sols = []  # list of all solutions for all combinations\n",
    "\n",
    "for s1, s2 in combinations(seqlist_WW, 2):\n",
    "    nw = NeedlemanWunsch(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=8,\n",
    "        E=2,\n",
    "        submat=pam60,\n",
    "        k=5\n",
    "    )\n",
    "    nw.run()\n",
    "    nw_sols.extend(nw.solutions) # add solutions to list of total solutions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Observations on error parameters\n",
    "In test3 and test4 which both use the same PAM60 matrix, we can observe that having error parameters set to (I=12,E=2) produces 21 solutions which is less than 30, produced by having error parameters set to (I=8,E=2).<br>\n",
    "The reason is that with greater error parameters the algorithm is more *strict* when penalizing gaps.\n",
    "Gaps must be added meticulously, forcing 2 sequences to add gaps won't reflect the reality.\n",
    "If the gap penalty is set high, then fewer gaps will be inserted in the alignment and therefore, fewer alignments are considered valid.\n",
    "\n",
    "### Observations on substitution matrices \n",
    "We can observe that for identical error parameters (I=12,E=2) and (I=8,E=2), using the PAM60 matrix produces different results than using the PAM120 matrix in tests 1,3 and 2,4.<br>\n",
    "In fact, the PAM60 matrix represents mutations with an evolutionary distance of 60, which is lower than 120, the evolutionary distance of PAM120. We can thus suppose that some of the solutions produced with PAM60, in this case, are acceptable on an evolutionary distance of 60 but not 120.\n",
    "\n",
    "### Comparison with LALIGN\n",
    "For global alignment, LALIGN only produces only one best alignment.\n",
    "\n",
    "In test1, the solutions produced by the implemented Needleman-Wunsch algorithm are all equal to solutions produced by LALIGN:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<table>\n",
    "<caption>Result comparison with LALIGN</caption>\n",
    "<tr>\n",
    "    <th>sequences</th>\n",
    "    <th>Our implementation</th>\n",
    "    <th>LALIGN</th>\n",
    "    <th>Identical</th>\n",
    "</tr>\n",
    "\n",
    "\n",
    "<tr>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP</td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "                \n",
    " :::.:::  .   :. :..:: .. : :. :  <br>\n",
    " \n",
    "SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP </td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "        :::.:::  .   :. :..:: .. : :. :  <br>\n",
    "        SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "            \n",
    "            <tr>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL</td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    " :::.:::    ..:. ...::  . : :.::: <br>\n",
    "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL </td>\n",
    "                <td>VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "        :::.:::    ..:. ...::  . : :.::: <br>\n",
    "        GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG</td>\n",
    "                <td> VPLPAGWEMAKT-SSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "  ::.:::. .. :::. :..::: ... :. :  <br>\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG </td>\n",
    "                <td>VPLPAGWEMAKT-SSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "         ::.:::. .. :::. :..::: ... :. :  <br>\n",
    " EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG </td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "    . :   :...:. :. :   . .::. :  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD </td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "           . :   :...:. :. :   . .::. :  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    ".     :   :..::. :. :   . . :  :. <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td> VPLPAGWEMAKTSSGQRYFLNHIDQTTTWQDPRK<br>\n",
    "       .     :   :..::. :. :   . . :  :. <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
    "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL</td>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    ".::::::::: .. ::..:.::. .::::  :  <br>\n",
    "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL </td>\n",
    "                <td>SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    "       .::::::::: .. ::..:.::. .::::  :  <br>\n",
    "GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG</td>\n",
    "                <td> SPLPPGWEERQD-ILGRTYYVNHESRRTQWKRPTP<br>\n",
    "  :::::: : .   ::.:: :: .  .:: ::. <br>\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG </td>\n",
    "                <td>SPLPPGWEERQDIL-GRTYYVNHESRRTQWKRPTP<br>\n",
    "         :::::: : .   ::.:: :: .  .:: ::. <br>\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    ":   . : :...  ::::: : :.... : .:  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD </td>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    "       :   . : :...  ::::: : :.... : .:  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD </td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    "      : : ..  :. :: : ... ..: .:  <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td> SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP<br>\n",
    "             : : ..  :. :: : ... ..: .:  <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG</td>\n",
    "                <td> GPLPPGWEER-THTDGRIFYINHNIKRTQWEDPRL<br>\n",
    "  :::::: : ....::..:.:: .. .::: :  <br>\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG </td>\n",
    "                <td>GPLPPGWEER-THTDGRIFYINHNIKRTQWEDPRL<br>\n",
    "         :::::: : ....::..:.:: .. .::: :  <br>\n",
    "EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL<br>\n",
    ".   . : :.   :::..: : ..:.. :: :  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD </td>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL<br>\n",
    "       .   . : :.   :::..: : ..:.. :: :  <br>\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL<br>\n",
    "      : :    .:. .: : ..: ..:. :. <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td> GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL<br>\n",
    "             : :    .:. .: : ..: ..:. :. <br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
    "SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG<br>\n",
    "    . :  ..: ..::.::.:  :. : ::.:..<br>\n",
    "SGAKSMWTEHKS-PDGRTYYYNTETKQSTWEKPDD </td>\n",
    "                <td>EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG<br>\n",
    "           . :  ..: ..::.::.:  :. : ::.:..<br>\n",
    "SGAKSMWTEHKS-PDGRTYYYNTETKQSTWEKPDD</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG<br>\n",
    "      : :    .::. ::.:  :..:.:..:  <br>\n",
    "LLSKCPW-KEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td> EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG<br>\n",
    "             : :    .::. ::.:  :..:.:..:  <br>\n",
    "LLSKCPW-KEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "<tr>\n",
    "                <td> SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD<br>\n",
    "  .:. : : :: .:. ::::..::.: :.:: .<br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE </td>\n",
    "                <td>SGAKSMWTEHKSPDGRTYYYNTETKQSTWEKPDD<br>\n",
    "         .:. : : :: .:. ::::..::.: :.:: .<br>\n",
    "LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE</td>\n",
    "                <td> True </td>\n",
    "            </tr>\n",
    "\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In some cases, the solution produced by LALIGN differs from our implementation:\n",
    "For instance let us consider the algorithm configured with PAM120, $I=8$ and $E=2$ running on sequences **EKLPPGWEKRMSRSSGRVYYFNHITNASQWERPSG** and  **LLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE**.\n",
    "\n",
    "LALIGN' solution:\n",
    "<img src=\"img/lalign2_global.png\">\n",
    "\n",
    "Our (best) solution:\n",
    "```\n",
    "EKLP--PGWEKRMSRSSGRVYYFNHITNASQWERPSG\n",
    "  :.  : : :    .::. ::.:  :..:.:..:  \n",
    "-LLSKCP-W-KEYKSDSGKPYYYNSQTKESRWAKPKE\n",
    "Identity : 35.14 %\n",
    "Length : 37\n",
    "```\n",
    "\n",
    "It is a difficult task to reproduce the same results as LALIGN because of some implementation choices not strictly defined by the original algorithm.\n",
    "We mostly observe differences with LALIGN when comparing alignments having gaps.\n",
    "We can hypothesize that LALIGN favors alignments with minimum gaps.\n",
    "\n",
    "### Statistical analysis\n",
    "\n",
    "In order to detect which sequences stand out, we decided to rely on statistical analysis.<br>\n",
    "We first calculate the average and standard deviation of the identity and similarity of all solutions produced in test1.\n",
    "\n",
    "Then, we compute solutions that have an identity level and/or similarity higher than 1.5 standard deviation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_stats(val_list):\n",
    "    \"\"\"Given a list of values returns the arithmetic mean and standard deviation\"\"\"\n",
    "    avg = sum(val_list)/len(val_list)\n",
    "    std_dev = (sum([(val-avg)**2 for val in val_list])/len(val_list))**0.5\n",
    "    return avg, std_dev\n",
    "\n",
    "def get_sd_standout(val_list, avg, sd, factor, func_getter):\n",
    "    \"\"\"\n",
    "    given a list of objects val_list, returns a list of objects\n",
    "    where func_getter(object) is higher by factor*standard deviation\n",
    "    \"\"\"\n",
    "    return [sol for sol in nw_sols if func_getter(sol)>=(avg+factor*sd)]\n",
    "\n",
    "# identity statistics\n",
    "avg_identity, sd_identity = get_stats([sol.identity for sol in nw_sols])\n",
    "# similarity statistics\n",
    "avg_similarity, sd_similarity = get_stats([sol.similarity for sol in nw_sols])\n",
    "\n",
    "print('Average identity: %.2f %%, SD identity: %.2f %%' % (avg_identity*100, sd_identity*100))\n",
    "print('Average similarity:%.2f %%, SD identity: %.2f %%'  %(avg_similarity*100, sd_similarity*100))\n",
    "\n",
    "# solutions with more than 2SDs, highly identical or similar\n",
    "FACTOR=2\n",
    "h_identicals = get_sd_standout(nw_sols, avg_identity, sd_identity, FACTOR, lambda x: x.identity)\n",
    "h_similars = get_sd_standout(nw_sols, avg_similarity, sd_similarity, FACTOR, lambda x: x.similarity)\n",
    "\n",
    "# intersection\n",
    "FACTOR=1.5\n",
    "identicals = get_sd_standout(nw_sols, avg_identity, sd_identity, FACTOR, lambda x: x.identity)\n",
    "similars = get_sd_standout(nw_sols, avg_similarity, sd_similarity, FACTOR, lambda x: x.similarity)\n",
    "id_and_sim = set(identicals).intersection(set(similars))\n",
    "\n",
    "for sol in h_identicals:\n",
    "    print('highly identical:', sol.s1, ' and ', sol.s2)\n",
    "    \n",
    "for sol in h_similars:\n",
    "    print('highly similar:', sol.s1, ' and ', sol.s2)\n",
    "    \n",
    "for sol in id_and_sim:\n",
    "    print('similar and identical:', sol.s1, ' and ', sol.s2)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We observe that **SPLPPGWEERQDILGRTYYVNHESRRTQWKRPTP**  and  **GPLPPGWEERTHTDGRIFYINHNIKRTQWEDPRL** are more similar and identical than any other pair of sequences. \n",
    "Their identity level is very high (> 2 SD) and their similarity level is high (> 1.5 SD).\n",
    "Therefore we suppose that they are related.\n",
    "\n",
    "Consulting Uniprot with the code name `P46934` in the FASTA file, it appears that the 2 sequences are segments from the <u>same</u> protein E3 ubiquitin-protein ligase NEDD4 belonging to Homo Sapiens (humans).\n",
    "<img src=\"img/uniprot_global.png\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local alignment \n",
    "\n",
    "For each pair of sequences in the `protein-sequence.fasta` file (*15 sequences in total*), we compute up to $l=3$ solutions using the Smith-Waterman algorithm with varying parameters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 5\n",
    "# Substitution matrix : BLOSUM62\n",
    "# I: 12\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: 45\n",
    "\n",
    "for s1, s2 in combinations(seqlist_protein, 2):\n",
    "    SmithWaterman(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=12,\n",
    "        E=2,\n",
    "        submat=blosum62,\n",
    "        l=3\n",
    "    ).run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### test 6\n",
    "# Substitution matrix : BLOSUM80\n",
    "# I: 12\n",
    "# E: 2\n",
    "# \n",
    "# Total number of solutions: \n",
    "for s1, s2 in combinations(seqlist_protein, 2):\n",
    "    SmithWaterman(\n",
    "        seq1=s1,\n",
    "        seq2=s2,\n",
    "        I=12,\n",
    "        E=2,\n",
    "        submat=blosum80,\n",
    "        l=3\n",
    "    ).run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Observations on substitution matrices \n",
    "We can observe that for identical error parameters (I=12,E=2), although the number of solutions is equal to 45 for both ($l$ alignmnents per pair), the alignments produced while using BLOSUM62 and BLOSUM80 are different.\n",
    "The BLOSUM80 matrix is based on an 80% minimum identity threshold and thus should be more strict than BLOSUM62 which is based on a 62% threshold at accepting solutions.\n",
    "\n",
    "### Comparison with LALIGN\n",
    "Running our implementation of Smith-Waterman algorithm yields in some cases different outputs than LALIGN configured with the same parameters.\n",
    "\n",
    "\n",
    "For instance, let us consider the P46937 and O75400 protein sequences of the `protein-sequences.fasta` file.\n",
    "\n",
    "The solutions produced by our algorithm, using BLOSUM62 and (I=12,E=2) are:\n",
    "\n",
    "```\n",
    "# solutions :  3\n",
    "\n",
    "RQSSFEIPDDVPLPA-------GWEMAKTSSGQRYFLNHIDQTTTWQDPRK\n",
    ".::..: :::.  ::        :.  :..::. :. :   . . :  :..\n",
    "KQSTWEKPDDLKTPAEQLLSKCPWKEYKSDSGKPYYYNSQTKESRWAKPKE\n",
    "Identity : 31.37 %\n",
    "Length : 51\n",
    "\n",
    "\n",
    "MLSQMN-VTA-PTSPPVQQNMMNSASGPL-PDG-WEQAMTQDGEIYYINHKNKTTSWLDP\n",
    ":.:.:. ..  :. ::  ..:  .:.      . : .  . ::. :: : ..: ..:  :\n",
    "MMSHMSQASMQPALPPGVNSMDVAAGTASGAKSMWTEHKSPDGRTYYYNTETKQSTWEKP\n",
    "Identity : 28.33 %\n",
    "Length : 60\n",
    "\n",
    "\n",
    "WEQAMTQDGEIYYINHKNKTTSWLDPR\n",
    ":..  ...:. :: : ..: . :  :.\n",
    "WKEYKSDSGKPYYYNSQTKESRWAKPK\n",
    "Identity : 29.63 %\n",
    "Length : 27\n",
    "```\n",
    "\n",
    "Which is the approximately the same as the LALIGN solution:\n",
    "<img src=\"img/lalign_local_1.png\">\n",
    "\n",
    "Let us consider the case using the first and the fourth sequence of the `protein-sequences.fasta` file:\n",
    "\n",
    "```\n",
    "# solutions :  3\n",
    "\n",
    "DDVPLPAGWEMAKT-SSGQRYFLNHIDQTTTWQDP\n",
    ":.  :: :::   . :::. :..::: ... :. :\n",
    "DEEKLPPGWEKRMSRSSGRVYYFNHITNASQWERP\n",
    "Identity : 42.86 %\n",
    "Length : 35\n",
    "\n",
    "\n",
    "LPDGWEQAMTQ-DGEIYYINHKNKTTSWLDP\n",
    ":: :::. :.. .:..::.:: .....:  :\n",
    "LPPGWEKRMSRSSGRVYYFNHITNASQWERP\n",
    "Identity : 41.94 %\n",
    "Length : 31\n",
    "\n",
    "\n",
    "TGVVSGPAATPTAQHL\n",
    ":: .:::. : .. :.\n",
    "TGEMSGPVFTDSGIHI\n",
    "Identity : 43.75 %\n",
    "Length : 16\n",
    "```\n",
    "\n",
    "Whereas LALIGN' solution is considerably different:\n",
    "<img src=\"img/lalign_local_2.png\">\n",
    "\n",
    "\n",
    "A potential reason is that the Smith-Waterman algorithm we were asked to implement is simplified compared to the one of LALIGN, considering the course INFO-F-208 is an introduction to bioinformatics. Another possible reason is that LALIGN implements the Smith-Waterman algorithm with custom optimisations.\n",
    "\n",
    "\n",
    "### Interpretation\n",
    "Let us consider the alignment of P46934 and Q9HCE7 segments, provided in `protein-sequences.fasta` for which a solution is :\n",
    "```\n",
    "EPSPLPPGWEERQDILGRTYYVNHESRRTQWKRP\n",
    ":  ::::::: :. . :: :.:.: .: ::.  :\n",
    "ELGPLPPGWEVRSTVSGRIYFVDHNNRTTQFTDP\n",
    "Score : 109\n",
    "Identity : 52.94 %\n",
    "Similarity : 70.59 %\n",
    "Length : 34\n",
    "```\n",
    "We have chosen to investigate this particular alignment because of the relatively high identity and similarity level.\n",
    "\n",
    "Based on the complementary biological information provided by Uniprot, we can indeed deduce functional similarities between the 2 sequences.\n",
    "\n",
    "In the *Function* section of the page, we read that both the P46934 (E3 ubiquitin-protein ligase NEDD4) and Q9HCE7 (E3 ubiquitin-protein ligase SMURF1) essentially perform ubiquination functions.<br>\n",
    "Both appear to have the same catalytic activity (*S-ubiquitinyl-[E2 ubiquitin-conjugating enzyme]-L-cysteine + [acceptor protein]-L-lysine = [E2 ubiquitin-conjugating enzyme]-L-cysteine + N(6)-ubiquitinyl-[acceptor protein]-L-lysine*).<br>\n",
    "\n",
    "Similarly, examining the *Family & Domains* section of  the page, , both proteins possess the WW1, WW2 and HECT domains.<br>\n",
    "Therefore, we presume that the presence of these domains might be associated to the fact that they share similar functionality, and that the amino acid segments aligned in this example are probably part of one of these domains.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion\n",
    "\n",
    "In conclusion, we examine in this report protein sequence alignment using the Needleman-Wunsch and Smith-Waterman algorithms for global and local alignment.\n",
    "Key principles are explained and illustrated throughout the report.\n",
    "\n",
    "The algorithms are implemented and tested with various parameters in an attempt to make sense of different concepts.\n",
    "The results are analyzed, compared to LALIGN and discussed accordingly in an attempt to produce hypotheses about the outcome.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference\n",
    "- M.Zvelebil, J.O.Baum, (2008). *Understanding bioinformatics*\n",
    "- https://www.uniprot.org/\n",
    "- https://embnet.vital-it.ch/software/LALIGN_form.html\n",
    "- https://www.majordifferences.com/2014/02/difference-between-pam-and-blosum-matrix_1.html\n",
    "- http://www.life.illinois.edu/mcb/150/private/faq/pdf/1390.pdf\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
